// @solarwinds/xui/ng2/demo/src/components/app/app.component.ts
import { Component, Inject, OnInit, ViewEncapsulation } from "@angular/core";
import { Event, NavigationStart, Router } from "@angular/router";
import demoContent from "../../content";

// TODO: Re-implement demo layout using xui components
@Component({
    selector: "app",
    template: "<router-outlet></router-outlet>",
    encapsulation: ViewEncapsulation.None,
    styles: [require<string>("../../../src/styles.less")]
})
export class AppComponent implements OnInit {

    selectedContentName: string;

    content: any[];

    constructor(private router: Router) { }

    ngOnInit() {
        this.content = [
            {
                id: "components",
                name: "Components",
                articles: demoContent.filter((c: any) => c.category === "Components")
                    .sort((a1, a2) => a1.id.localeCompare(a2.id))
            }, {
                id: "styles",
                name: "Styles",
                articles: demoContent.filter((c: any) => c.category === "Styles")
                    .sort((a1, a2) => a1.id.localeCompare(a2.id))
            }, {
                id: "directives",
                name: "Directives",
                articles: demoContent.filter((c: any) => c.category === "Directives")
                    .sort((a1, a2) => a1.id.localeCompare(a2.id))
            }
        ];

        this.router.events.subscribe((event: Event) => {
            if (event instanceof NavigationStart) {
                let selectedContent = demoContent.find(c => c.url === event.url.replace("/", ""));
                this.selectedContentName = selectedContent ? selectedContent.name : "";
            }
        });
    }

}
