// @solarwinds/xui/sdk/spec/components/scrollbar/scrollbar.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class ScrollbarAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.hasScrollbar = (element) => {
            return element.getAttribute("clientHeight").then((clientHeight) => {
                return element.getAttribute("scrollHeight").then((scrollHeight) => {
                    return parseInt(scrollHeight, 10) > parseInt(clientHeight, 10);
                });
            });
        };
        this.scroll = (scrollbar, delta) => {
            return scrollbar.getLocation().then((location) => {
                return protractor_1.browser.actions()
                    .dragAndDrop(scrollbar.getWebElement(), { x: location.x + delta.x, y: location.y + delta.y })
                    .perform();
            });
        };
        this.scrollToElement = (target) => {
            return protractor_1.browser.executeScript("arguments[0].scrollIntoView();", target.getWebElement());
        };
        this.getScrollbarElement = (selector) => {
            return super.getElement().element(protractor_1.by.css(selector));
        };
        this.isTransparent = (color) => {
            return color === "rgba(0, 0, 0, 0)";
        };
    }
    static findIn(atomClass, parentElement, index) {
        if (!parentElement) {
            parentElement = protractor_1.browser.element(protractor_1.by.xpath("/"));
        }
        let componentRootElement;
        if (index !== undefined) {
            componentRootElement = parentElement.all(protractor_1.by.css("[xui-scrollbar]")).get(index);
        }
        else {
            componentRootElement = parentElement.element(protractor_1.by.css("[xui-scrollbar]"));
        }
        return new atomClass(componentRootElement);
    }
}
exports.ScrollbarAtom = ScrollbarAtom;
