// @solarwinds/xui/sdk/spec/components/checkbox/checkbox.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class CheckboxAtom extends uzi_1.Atom {
    constructor(element) {
        super(element);
        this.getInputElement = () => {
            return super.getElement().element(protractor_1.by.className("xui-checkbox-input"));
        };
        this.getLabel = () => {
            return super.getElement().element(protractor_1.by.className("xui-checkbox-label"));
        };
        this.getContent = () => {
            const element = super.getElement().element(protractor_1.by.className("xui-checkbox-transclude"));
            return element.getText();
        };
        this.getHelpHintText = () => {
            const element = super.getElement().element(protractor_1.by.className("xui-input-help"));
            return element.getText();
        };
        this.getValidationText = () => {
            const element = super.getElement().element(protractor_1.by.className("xui-validation"));
            return element.getText();
        };
        this.isInvalid = () => {
            return this.getInputElement().getAttribute("aria-invalid").then((value) => {
                return value === "true";
            });
        };
        this.isIndeterminate = () => {
            return this.getInputElement().getAttribute("indeterminate").then((value) => {
                return value === "true";
            });
        };
        this.isRequired = () => {
            return this.getInputElement().getAttribute("required").then((value) => {
                return value === "true";
            });
        };
        this.isChecked = () => {
            const cbInput = this.getInputElement();
            return cbInput.getAttribute("checked").then((value) => {
                return value === "true";
            });
        };
        this.isDisabled = () => {
            return this.getInputElement().isEnabled().then((value) => {
                return !value;
            });
        };
        this.click = () => {
            return this.getMark().click();
        };
    }
    getMark() {
        return this.getElement().element(protractor_1.by.className("xui-checkbox-mark"));
    }
}
CheckboxAtom.CSS_CLASS = "xui-checkbox";
exports.CheckboxAtom = CheckboxAtom;
