// @solarwinds/xui/ng2/src/components/menu/menu.component.ts
import { Component, Input, HostBinding, OnChanges, SimpleChanges } from "@angular/core";
import { IMenuItem } from "./menu-item";
/**
 * Examples: <example-url>./../../../demo/index.html#/menu</example-url><br />
 * "xui-menu" provides simple dropdown menu option with custom items:
 * "xui-header", "xui-divider", "xui-action", "xui-link", "xui-option"
 */
@Component({
    selector: "xui-menu",
    host: {
        "class": "xui-menu",
    },
    template: <string>require("./menu.component.html")
})
export class MenuComponent implements OnChanges {
    /**
     * sets title/name for menu button
     */
    @Input() public title: string;
    /**
     * sets size of menu button and caret icon
     * (take a look on xui-button possible sizes)
     */
    @Input() public size?: string;
    /**
     * sets styles for menu button
     * (take a look on xui-button possible styles)
     */
    @Input() public displayStyle?: string;
    /**
     * adds additional icon at the left of menu button
     * (take a look on xui-icon possible values)
     */
    @Input() public icon?: string;
    /**
     *Value to be used as menu items data.
     */
    @Input() public itemsSource?: IMenuItem[];

    @HostBinding("class.xui-menu--sm") public get smCssClass() {
        return this.size === "small";
    }

    @HostBinding("class.xui-menu--lg") public get lgCssClass() {
        return this.size === "large";
    }

    public iconSize = "";

    public iconColor = "";

    public ngOnChanges(changes: SimpleChanges): void {
        if (changes["size"]) {
            this.iconSize = this.size === "small" ? this.size : "";
        }

        if (changes["displayStyle"]) {
            this.iconColor = this.displayStyle === "primary" ? "white" : "";
        }
    }

}
