// @solarwinds/xui/ng2/src/components/menu/menu-option.component.ts
import { Component, Input, HostListener } from "@angular/core";

import { MenuItemComponent } from "./menu-item";

/**
 * Menu item component with check/uncheck option,
 * use (change) event to get state of this item
 */
@Component({
    selector: "xui-menu-option",
    template: `
        <div class="xui-menu-item__option">
            <xui-checkbox class="xui-menu-item__checkbox"
                [name]="name"
                [value]="value"
                [checked]="checked"
                [disabled]="disabled">

                <ng-content></ng-content>
            </xui-checkbox>
        </div>
    `
})
export class MenuOptionComponent extends MenuItemComponent {
    /**
     * Sets inner input "name" attribute
     */
    @Input() public name: string;
    /**
     * Sets inner input "value" attribute
     */
    @Input() public value: string;
    /**
     * Is needed to predefine item state, sets xui-checkbox [checked] property
     */
    @Input() public checked: boolean;

    @HostListener("click", ["$event"]) public stopPropagationOfClick(event: MouseEvent) {
        event.stopPropagation();
    }

    constructor() {
        super();

        this.name = "";
        this.value = "";
        this.checked = false;

        // Is needed to predefine item state, sets xui-checkbox [disabled] property
        this.disabled = false;
    }

}
