// @solarwinds/xui/ng2/demo/src/components/demo/edge-detection-service/edge-detection-service.example.component.ts
import {
    Component,
    ElementRef,
    Inject,
    ViewChild
} from "@angular/core";

import * as xui from "xui";
import { EdgeDetectionService, IEdgeDetectionResult } from "../../../../../src/services";

@Component({
    selector: "xui-edge-detection-service-example",
    template: require<string>("./edge-detection-service.example.component.html"),
    styles: [
        require<string>("./edge-detection-service.example.component.less")
    ]
})
export class EdgeDetectionServiceExampleComponent {
    @ViewChild("parent") private parentElement: ElementRef;
    @ViewChild("placementElement") private toBePlacedElement: ElementRef;

    public showPlaced = {};
    public showAligned = {};
    public deposit = {
        width: 100,
        height: 50
    };
    public parentComponent = {
        width: 300,
        height: 150
    };
    public addEdgeDefinerClass = false;

    public canBe: IEdgeDetectionResult = {
        placed: {
            top: false,
            right: false,
            bottom: false,
            left: false
        },
        aligned: {
            top: false,
            right: false,
            bottom: false,
            left: false
        }
    };

    constructor(private edgeDetectionService: EdgeDetectionService) {
    }

    public ngAfterViewInit(): void {
        this.update();
    }

    public update() {
        setTimeout(() => {
            const parent = this.parentElement.nativeElement;
            const basePointElement = parent.querySelector(".base-point-element");

            if (this.parentComponent.width < 300) {
                this.parentComponent.width = 300;
            }

            if (this.parentComponent.height < 50) {
                this.parentComponent.height = 50;
            }

            if (this.addEdgeDefinerClass) {
                parent.classList.add("xui-edge-definer");
            } else {
                parent.classList.remove("xui-edge-definer");
            }

            this.canBe = this.edgeDetectionService
                .canBe(basePointElement, this.toBePlacedElement.nativeElement);
        });
    }

    public switchCheckbox(): void {
        this.addEdgeDefinerClass = !this.addEdgeDefinerClass;
        this.update();
    }

    public changeText(path: string, value: any) {
        _.set(this, path, value);
        this.update();
    }
}
