// @solarwinds/xui/ng1/sdk/spec/components/globalKeys/globalKeys.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor = require("@solarwinds/uzi/protractor");
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class GlobalKeysAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.isXuiRevealActive = () => {
            return super.hasClass("xui-reveal");
        };
        this.focus = () => {
            return super.getElement().click();
        };
        this.toggleXuiReveal = () => {
            return super.getElement()
                .sendKeys(protractor.Key.chord(protractor.Key.CONTROL, protractor.Key.SHIFT, "`"));
        };
        this.getComponents = (name) => {
            return super.getElement().all(protractor_1.by.className(name));
        };
        this.hasRedOutline = (el) => {
            return el.getCssValue("outline-color")
                .then((val) => {
                return val === "rgba(255, 0, 0, 1)";
            });
        };
    }
}
exports.GlobalKeysAtom = GlobalKeysAtom;
