// @solarwinds/xui/ng1/sdk/spec/components/expander/expander.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class ExpanderAtom extends uzi_1.Atom {
    isExpanded() {
        return super.getElement()
            .element(protractor_1.by.css(".xui-expander-body.collapse.in"))
            .isPresent();
    }
    isContentDisplayed() {
        return super.getElement().element(protractor_1.by.css(".xui-expander-body")).isDisplayed();
    }
    isContentAttachedToDOM() {
        return super.getElement().element(protractor_1.by.css(".xui-expander-body div")).isPresent();
    }
    toggle() {
        return super.getElement().element(protractor_1.by.css(".xui-expander-heading-icon")).click();
    }
    getHeading() {
        return super.getElement().element(protractor_1.by.css(".xui-expander-heading-content"));
    }
    getHeadingText() {
        return super.getElement().element(protractor_1.by.css(".xui-expander-heading-content")).getText();
    }
}
ExpanderAtom.CSS_CLASS = "xui-expander";
exports.ExpanderAtom = ExpanderAtom;
