// @solarwinds/xui/ng1/src/components/chiclets/chiclets-directive.ts
/// <reference path="../../ref.d.ts" />

import ChicletsController from "./chiclets-controller";

/**
 * @ngdoc directive
 * @name xui.directive:xuiChiclets
 * @restrict E
 *
 * @description
 * Control used to list groups of selected options.
 *
 * @parameters
 * @param {boolean} [autoHide=true] Expression to determine if the component is hidden when 'itemsSource' is empty.
 * @param {string} [clearAllLinkText="Clear all"] Value to be shown as clear all link text.
 * @param {function} itemLabelFn (itemId:string, groupId:string)=>string function returning item/group label based on
 * given IDs.
 * @param {string[] | IMap<any[]>} itemsSource Either flat (string[]) or grouped (IMap<any[]>) structure.
 * If grouped, object properties are parsed as groups.
 * @param {string} [orientation="horizontal"] Can be set to 'vertical' to switch chiclets to vertical list mode.
 * @param {boolean} [allowClearAll=true] Flag for 'clear all' link visibility.
 * @param {string} title Value to be shown as chiclets title.
 * @param {boolean} [isDraggable=false] Changes appearance of chiclets. PLEASE DO NOT USE, THIS FEATURE IS UNDER
 * CONSTRUCTION. For more info see UIF-4341.
 *
 * @example
 *    <example module="xui">
 *        <file src="src/components/chiclets/docs/chiclets-examples.html" name="index.html"></file>
 *        <file src="src/components/chiclets/docs/chiclets-examples.js" name="script.js"></file>
 *        <file src="src/components/chiclets/docs/chiclets-examples.css" name="script.css"></file>
 *    </example>
 **/

class Chiclets implements ng.IDirective {
    public restrict = "E";
    public templateUrl = "xui/components/chiclets/chiclets-directive.html";
    public replace = true;
    public transclude = false;
    public scope = {};
    public bindToController = {
        allowClearAll: "=?",
        autoHide: "=?",
        clearAllLinkText: "@?",
        itemLabelFn: "&?",
        itemsSource: "=",
        isDraggable: "<?",
        orientation: "@?",
        title: "@?",
        onRemove: "&?"
    };
    public controller = ChicletsController;
    public controllerAs = "vm";
}

export default Chiclets;
