// @solarwinds/xui/sdk/spec/components/panel/panel.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class PanelAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.toggleButton = () => {
            return super.getElement().element(protractor_1.by.className("xui-panel__heading-btn"));
        };
        this.icon = () => {
            return super.getElement().element(protractor_1.by.className("xui-panel__heading-icon"));
        };
        this.getHeaderIconCounter = (state) => {
            const selector = `${PanelAtom.ICON_CSS}--${state}`;
            const headerIconElement = super.getElement().element(protractor_1.by.className(selector));
            return headerIconElement.getAttribute("counter");
        };
        this.isCollapsed = () => {
            return super.hasClass(PanelAtom.COLLAPSED_CSS);
        };
        this.isExpanded = () => {
            return super.hasClass(PanelAtom.COLLAPSED_CSS).then((value) => {
                return !value;
            });
        };
        this.iconShowsCollapsed = () => {
            return uzi_1.Atom.hasClass(this.icon(), PanelAtom.COLLAPSED_ICON_CSS);
        };
        this.iconShowsExpanded = () => {
            return uzi_1.Atom.hasClass(this.icon(), PanelAtom.EXPANDED_ICON_CSS);
        };
        this.headingText = () => {
            return super.getElement().element(protractor_1.by.css(".xui-panel__heading-content h4")).getText();
        };
    }
}
PanelAtom.CSS_CLASS = "xui-panel";
PanelAtom.COLLAPSED_CSS = "xui-is-collapsed";
PanelAtom.EXPANDED_ICON_CSS = "xui-icon-double-caret-left";
PanelAtom.COLLAPSED_ICON_CSS = "xui-icon-double-caret-right";
PanelAtom.ICON_CSS = "xui-panel__heading-icon";
exports.PanelAtom = PanelAtom;
