// @solarwinds/xui/sdk/spec/components/ellipsis/ellipsis.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
const specHelpers_1 = require("../specHelpers");
class EllipsisAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.moveMouseToTarget = (atom) => {
            return specHelpers_1.default.mouseOver(atom);
        };
    }
    getTitle() {
        return this.getElement()
            .element(protractor_1.by.className("xui-ellipsis-inner__text")).getText();
    }
    getEllipsedAttribute(attrName) {
        return this.getElement().element(protractor_1.by.className("xui-ellipsis-inner")).getAttribute(attrName);
    }
    isTooltipDisplayed() {
        return this.getTooltip().isPresent();
    }
    getTooltipAttribute(attribute) {
        return this.getTooltip().getAttribute(attribute);
    }
    getTooltip() {
        return protractor_1.browser.element(protractor_1.by.className("ellipsis-tooltip"));
    }
}
EllipsisAtom.CSS_CLASS = "xui-ellipsis";
exports.EllipsisAtom = EllipsisAtom;
