// @solarwinds/xui/sdk/spec/components/radio/radio.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class RadioAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.isRadioSelected = (radio) => {
            let inputElement = this.getRadioInputElement(radio);
            return inputElement.isSelected();
        };
        this.isRadioEnabled = (radio) => {
            let inputElement = this.getRadioInputElement(radio);
            return inputElement.isEnabled();
        };
        this.getRadioText = (radio) => {
            return radio.getText();
        };
        this.getValue = (radio) => {
            let inputElement = this.getRadioInputElement(radio);
            return inputElement.getAttribute("value");
        };
        this.clickOnRadio = (radio) => {
            // Click on label, not the input element
            return radio.element(protractor_1.by.className("xui-radio-label")).click();
        };
        this.findRadioByName = (radioList, radioName) => {
            return radioList.filter((radio) => {
                return this.getRadioText(radio).then((text) => {
                    return text.toLowerCase() === radioName.toLowerCase();
                });
            }).first();
        };
        this.getRadioInputElement = (radio) => {
            return radio.element(protractor_1.by.className("xui-radio-input"));
        };
    }
}
RadioAtom.CSS_CLASS = "xui-radio";
exports.RadioAtom = RadioAtom;
