// @solarwinds/xui/sdk/spec/components/selector/selector.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class SelectorAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.getCheckboxElement = () => {
            return super.getElement().element(protractor_1.by.className("xui-checkbox-input"));
        };
        this.getCheckboxButton = () => {
            return super.getElement().element(protractor_1.by.className("xui-selector__action"));
        };
        this.getDropdownButton = () => {
            return super.getElement().element(protractor_1.by.className("xui-selector__toggle"));
        };
        this.clickOnDropdownItem = (index) => {
            this.getDropdownButton().click();
            this.getElementFromListByIndex(index).click();
        };
        this.isChecked = () => {
            const cbInput = this.getCheckboxElement();
            return cbInput.getAttribute("checked").then((value) => {
                return value === "true";
            });
        };
        this.isIndeterminate = () => {
            return this.getCheckboxElement().getAttribute("indeterminate").then((value) => {
                return value === "true";
            });
        };
    }
    getElementsList() {
        return super.getElement().element(protractor_1.by.className("xui-selector__menu")).getWebElement();
    }
    getElementFromListByIndex(index) {
        return super.getElement().all(protractor_1.by.className("xui-selector__menuitem"))
            .get(index).getWebElement();
    }
}
SelectorAtom.CSS_CLASS = "xui-selector";
exports.SelectorAtom = SelectorAtom;
