// @solarwinds/xui/ng1/src/common/constants.ts
/// <reference path="../ref.d.ts" />

import Module from "./module";
import IDragScrollConstants = xui.IDragScrollConstants;
import IUnitConversionConstants = xui.IUnitConversionConstants;

/**
 * XUI constants
 */
export class Constants {

    public KEY_CODE = {
        BACKSPACE: 8,
        TAB: 9,
        ENTER: 13,
        ESCAPE: 27,
        SPACE: 32,
        PAGE_UP: 33,
        PAGE_DOWN: 34,
        END: 35,
        HOME: 36,
        LEFT_ARROW: 37,
        UP_ARROW: 38,
        RIGHT_ARROW: 39,
        DOWN_ARROW: 40,
        DELETE: 46,
        X: 88,
        ASTERISK: 106,
        PLUS: 107,
        MINUS: 109,
        TILDE: 192
    };

    public alertDurationMs = 5000;

    public edgeDefinerSelector = ".xui-edge-definer";
}

const ScrollConstants:IDragScrollConstants  = {
    verticalScrollPercentage: 30,
    horizontalScrollPercentage: 30,
    checkIntervalInMs: 16,
    accelerationFactor: 0.5,
    dragOverThrottleInMs : 100,
    adornerUpdateThrottleInMs: 30
};

const UnitConversionConstants: IUnitConversionConstants  = {
    bytes: ["B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"],
    bitsPerSecond: ["bps", "kbps", "Mbps", "Gbps", "Tbps", "Pbps", "Ebps", "Zbps", "Ybps"],
    bytesPerSecond: ["Bps", "kBps", "MBps", "GBps", "TBps", "PBps", "EBps", "ZBps", "YBps"],
    hertz: ["Hz", "kHz", "MHz", "GHz", "THz", "PHz", "EHz", "ZHz", "YHz"]
};

export default (module:Module)=> {
    module.service("xuiConstants", Constants);
    module.app()
        .constant("xuiDragScrollConstants", ScrollConstants)
        .constant("xuiUnitConversionConstants", UnitConversionConstants);
};
