// @solarwinds/xui/ng2/demo/src/components/demo/listview/listview.example.component.ts
import { Component } from "@angular/core";
import {Logger} from "angular2-logger/app/core/logger";
import { IEmptyData } from "../../../../../src/index";

@Component({
    selector: "listview-example",
    template: require<string>("./listview.example.component.html")
})
export class ListViewExampleComponent {
    public emptySource: any[] = [];
    public selectedMode: string;
    public modes = ["single", "multi", "none"];

    public colors = [
        {color: "blue"},
        {color: "green"},
        {color: "yellow"},
        {color: "cyan "},
        {color: "magenta"},
        {color: "black"}
    ];
      
    public colorsWithIsDisabledProperty = [
        { color: "blue", disabled: true },
        { color: "green", disabled: false },
        { color: "yellow" },
        { color: "cyan " },
        { color: "magenta", disabled: true },
        { color: "black" }
    ];

    public people = [
        {name: "Jo Smith", level: "bronze", status:"active"},
        {name: "Claire Rogan", level: "gold", status:"active"},
        {name: "Declan McGregor", level: "platinum", status:"active"},
        {name: "Fergus O'Brien", level: "bronze", status:"inactive"},
        {name: "Catriona Kildare", level: "gold", status:"active"}
    ];

    public selectedPeople = [this.people[2]];
    public selectedColors = [this.colors[1]];
    public narrowRowSelectedColors = [this.colors[2], this.colors[5]];
    public selectedColorsWithIsDisabledProperty = [this.colorsWithIsDisabledProperty[0], 
                                                   this.colorsWithIsDisabledProperty[5]];
    
    public onPeopleSelectionChange(selection: any) {
        this.selectedPeople = selection;
    }

    public onColorSelectionChange(selection: any) {
        this.selectedColors = selection;
    }

    public onNarrowRowSelectionChange(selection: any) {
        this.narrowRowSelectedColors = selection;
    }
    
    private customTemplateViewModel = {
        image: "http://oriondemo.solarwinds.com/ui/modules/platform/images/404-Robot.svg",
        searchTerm: "Node134"
    };
    
    private customTemplate =
        `<div class="empty-container  sample-custom-template"> 
            <img src="${this.customTemplateViewModel.image}"
                 alt="No items in the list" class="empty-pic"/> 
            <p>
                 No result found for 
                 <span class="sample-custom-template_bound-value">${this.customTemplateViewModel.searchTerm}</span>
            </p> 
        </div>`;

    public customTemplateData = {
        template: this.customTemplate
    } as IEmptyData;
}
