// @solarwinds/xui/ng2/demo/src/components/demo/popup/popup.example.component.ts
import { Component } from "@angular/core";

@Component({
    selector: "popup-example",
    template: <string>require("./popup.example.component.html")
})
export class PopupExampleComponent {

    public barPopupOpened = false;

    public colorsPopupdirectionTop = false;

    public fooPopupDirectionTop = false;

    public colors = ["red", "green", "blue"];

    public selectedColors = new Set<string>();

    public handleChecked(event: any, fieldNameToChange: string): void {
        const target: { checked: boolean, value: string } = event.target;

        switch (fieldNameToChange) {
            case "barPopupOpened":
                this.barPopupOpened = target.checked;
                break;

            case "colorsPopupdirectionTop":
                this.colorsPopupdirectionTop = target.checked;
                break;

            case "fooPopupDirectionTop":
                this.fooPopupDirectionTop = target.checked;
                break;

            case "selectedColors":
                if (target.checked) {
                    this.selectedColors.add(target.value);
                } else {
                    this.selectedColors.delete(target.value);
                }
                break;
        }

    }

}
