// @solarwinds/xui/ng1/src/components/expander/expander-directive.ts
import IAttributes = angular.IAttributes;
import IAugmentedJQuery = angular.IAugmentedJQuery;
import IScope = angular.IScope;

import { ExpanderController } from "./expander-controller";

/**
 * @ngdoc directive
 * @name xui.directive:xuiExpander
 * @restrict E
 *
 * @description
 * XUI Expander component.
 *
 * @parameters
 * @param {boolean} [is-open=false] Whether expander should be expanded by default.
 * @param {string} heading Heading text.
 * @param {boolean} [is-disabled=false] Whether the expander should be disabled.
 * @param {function} [on-status-changed=null] If this function is set, receives IExpanderChangeArgs object in case of
 * opening or closing the expander.
 *
 * @example
 * <example module="xui">
 * <file src="src/components/expander/docs/expander-examples.html" name="index.html"></file>
 * <file src="src/components/expander/docs/expander-examples.css" name="index.css"></file>
 * <file src="src/components/expander/docs/expander-examples.js" name="app.js"></file>
 *</example>
 */

export class Expander implements ng.IDirective {
    public restrict = "E";
    public replace = true;
    public transclude:{[slot:string]:string} = {
        "header": "?xuiExpanderHeading"
    };
    public scope = {};
    public bindToController = {
        heading: "@?",
        isOpen: "=?",
        isDisabled: "=?",
        icon: "@?",
        onStatusChanged: "&?"
    };
    public templateUrl = "xui/components/expander/expander-directive.html";
    public controller = "xuiExpanderController";
    public controllerAs = "vm";
}
