// @solarwinds/xui/ng1/sdk/spec/components/pageContent/pageContent.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
const index_atom_1 = require("../index.atom");
class PageContentAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.getTitle = () => {
            return super.getElement().element(protractor_1.by.className("xui-page-content__header-title")).getText();
        };
        this.isIconDisplayed = (iconName) => {
            return super.getElement().all(protractor_1.by.css(".xui-page-content__header-link i")).get(0)
                .getAttribute("icon")
                .then((attrVal) => {
                return attrVal === iconName;
            });
        };
        this.getLink = () => {
            return super.getElement().all(protractor_1.by.css(".xui-page-content__header-link a")).get(0);
        };
        this.getLinkText = () => {
            return this.getLink().getText();
        };
        this.getLinkUrl = () => {
            return this.getLink().getAttribute("href");
        };
        this.getLinkTarget = () => {
            return this.getLink().getAttribute("target");
        };
        this.isBusyIndicatorDisplayed = (elId) => {
            return uzi_1.Atom.find(index_atom_1.BusyAtom, elId).isDisplayed();
        };
        this.getBusyMessage = () => {
            const busy = uzi_1.Atom.findIn(index_atom_1.BusyAtom, super.getElement());
            return busy.getProgress().getMessage();
        };
        this.getHeader = () => {
            return super.getElement().element(protractor_1.by.css("[ng-transclude='header']"));
        };
        this.isDefaultHeader = () => {
            return this.getHeader()
                .element(protractor_1.by.className("xui-page-content__header"))
                .isPresent()
                .then((isPresent) => isPresent);
        };
        this.getFooter = (elId) => {
            return super.getElement()
                .element(protractor_1.by.css("[ng-transclude='header']"))
                .element(protractor_1.by.id(elId));
        };
        this.getBreadcrumbs = () => {
            return super.getElement().element(protractor_1.by.css("[ng-transclude='breadcrumbs']"));
        };
        this.isBodyScrollable = () => {
            const body = super.getElement().element(protractor_1.by.className("xui-page-content__body"));
            const scrollbarAtom = uzi_1.Atom.findIn(index_atom_1.ScrollbarAtom, body);
            return scrollbarAtom.hasScrollbar(body);
        };
    }
}
PageContentAtom.CSS_CLASS = "xui-page-content";
exports.PageContentAtom = PageContentAtom;
