// @solarwinds/xui/ng2/src/module.ts
import { ModuleWithProviders, NgModule, Provider } from "@angular/core";
import { FormsModule } from "@angular/forms";
import { BrowserModule } from "@angular/platform-browser";
import { BrowserAnimationsModule } from "@angular/platform-browser/animations";

import { Logger } from "angular2-logger/app/core/logger";

import * as components from "./components";
import * as directives from "./directives";
import * as services from "./services";
import * as pipes from "./pipes";
import * as constants from "./constants";

@NgModule({
    imports: [
        BrowserModule,
        BrowserAnimationsModule,
        FormsModule
    ],
    declarations: [
        components.CheckboxComponent,
        components.CheckboxGroupComponent,
        components.ContentComponent,
        components.CodeComponent,
        components.CodeItemComponent,
        components.ExpanderComponent,
        components.HelpHintComponent,
        components.MessageComponent,
        components.DividerComponent,
        components.IconComponent,
        components.SearchComponent,
        components.SidebarComponent,
        components.SidebarContainerComponent,
        components.SpinnerComponent,
        components.SwitchComponent,
        components.ButtonComponent,
        components.PageContentComponent,
        components.PagerComponent,
        components.ProgressComponent,
        components.PanelComponent,
        components.BusyComponent,
        components.RadioComponent,
        components.RadioGroupComponent,
        components.ImageComponent,
        components.ListViewComponent,
        components.ListItemComponent,
        components.EmptyComponent,
        components.TextboxComponent,
        components.TreeviewComponent,
        components.TreeviewItemComponent,
        components.WizardStepComponent,
        components.MenuActionComponent,
        components.MenuOptionComponent,
        components.MenuHeaderComponent,
        components.MenuLinkComponent,
        components.MenuComponent,
        components.ToastComponent,
        components.IfShowComponent,
        components.PopupComponent,
        components.PopupToggleDirective,
        components.SelectorComponent,
        components.EllipsisComponent,
        components.FileUploadComponent,
        components.GridComponent,
        directives.ScrollShadowsDirective,
        directives.SetFocusDirective,
        directives.ValidationHandlerAsyncDirective,
        directives.ValidationHandlerDirective,
        directives.ClipboardDirective,
        directives.ToastDirective,
        directives.ResizeDirective,
        directives.DraggableDirective,
        directives.DroppableDirective,
        directives.TooltipDirective,
        directives.TooltipContent,
        pipes.HighlightPipe,
        pipes.UnitConversionPipe
    ],
    exports: [
        components.CheckboxComponent,
        components.CheckboxGroupComponent,
        components.ContentComponent,
        components.CodeComponent,
        components.CodeItemComponent,
        components.ExpanderComponent,
        components.HelpHintComponent,
        components.MessageComponent,
        components.DividerComponent,
        components.IconComponent,
        components.SearchComponent,
        components.SidebarComponent,
        components.SidebarContainerComponent,
        components.SpinnerComponent,
        components.SwitchComponent,
        components.ButtonComponent,
        components.PageContentComponent,
        components.PagerComponent,
        components.ProgressComponent,
        components.PanelComponent,
        components.BusyComponent,
        components.RadioComponent,
        components.RadioGroupComponent,
        components.ImageComponent,
        components.ListViewComponent,
        components.ListItemComponent,
        components.EmptyComponent,
        components.TextboxComponent,
        components.TreeviewComponent,
        components.WizardStepComponent,
        components.MenuActionComponent,
        components.MenuOptionComponent,
        components.MenuHeaderComponent,
        components.MenuLinkComponent,
        components.MenuComponent,
        components.IfShowComponent,
        components.PopupComponent,
        components.PopupToggleDirective,
        components.SelectorComponent,
        components.EllipsisComponent,
        components.GridComponent,
        components.FileUploadComponent,
        directives.ScrollShadowsDirective,
        directives.SetFocusDirective,
        directives.ValidationHandlerAsyncDirective,
        directives.ValidationHandlerDirective,
        directives.ClipboardDirective,
        directives.ToastDirective,
        directives.ResizeDirective,
        directives.DraggableDirective,
        directives.DroppableDirective,
        directives.TooltipDirective,
        directives.TooltipContent,
        pipes.HighlightPipe,
        pipes.UnitConversionPipe
    ],
    entryComponents: [
        components.ToastComponent,
        directives.TooltipContent
    ]
})
export class XuiModule {
    static forRoot(): ModuleWithProviders {
        return {
            ngModule: XuiModule,
            providers: [
                Logger,
                services.IconService,
                services.UnitConversionService,
                services.SwUtilService,
                services.TimeframeService,
                services.TransientCacheService,
                services.SearchService,
                services.NotificationService,
                services.ToastService,
                services.ToastContainerService,
                services.EventPropagationService,
                services.EdgeDetectionService,
                { provide: constants.UnitConversionToken, useValue: constants.UnitConversionConstants } as Provider,
                { provide: constants.ImagesPresetToken, useValue: constants.IMAGES_PRESET } as Provider,
                { provide: constants.EmptyPresetToken, useValue: constants.EMPTY_PRESET } as Provider
            ]
        };
    }
}
