// @solarwinds/xui/ng2/spec/components/progress/progress.atom.ts
import { by, promise } from "@solarwinds/uzi/protractor";
import { Atom } from "@solarwinds/uzi";

export class ProgressAtom extends Atom {

    public static CSS_CLASS = "xui-progress";

    public async getWidth(): Promise<number> {
        return Number((await this.getElement().getCssValue("width")).replace("px", ""));
    }

    public async getActualHeight(): Promise<number> {
        return (await this.getElement().getSize()).height;
    }

    public isDisplayed(): promise.Promise<boolean> {
        return this.getElement().isDisplayed();
    }

    public isProgressBarDisplayed(): promise.Promise<boolean> {
        return this.getElement().element(by.className("xui-progress__bar")).isDisplayed();
    }

    public getMessage(): promise.Promise<string> {
        return this.getElement().element(by.className("xui-progress__message")).getText();
    }

    public canCancel(): promise.Promise<boolean> {
        return this.getElement().element(by.className("xui-progress__cancel")).isPresent();
    }

    public cancelProgress(): void {
        this.getElement().element(by.className("xui-progress__cancel")).click();
    }

}
