// @solarwinds/xui/ng1/sdk/spec/components/button/button.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
const index_atom_1 = require("../index.atom");
class ButtonAtom extends uzi_1.Atom {
    click() {
        return super.getElement().click();
    }
    isVisible() {
        return super.getElement().isDisplayed();
    }
    isDisabled() {
        return super.getElement().isEnabled().then((value) => {
            return !value;
        });
    }
    isIconShown() {
        const icons = super.getElement().all(protractor_1.by.className("xui-icon"));
        return icons.count().then((count) => {
            return count > 0;
        });
    }
    isBusy() {
        return super.getElement().getAttribute("is-busy").then((attribute) => {
            /* tslint:disable:triple-equals */
            //we want type coercion here
            return (attribute == "true");
        });
    }
    getTextColor() {
        return super.getElement().getCssValue("color");
    }
    getBackgroundColor() {
        return super.getElement().getCssValue("background-color");
    }
    getBorderStyle() {
        return super.getElement().getCssValue("border-color");
    }
    getIcon() {
        return this.isIconShown().then((isShown) => {
            if (isShown) {
                return new index_atom_1.IconAtom(super.getElement().all(protractor_1.by.className("xui-icon")).first());
            }
            else {
                return;
            }
        });
    }
    getText() {
        return super.getElement().element(protractor_1.by.className("xui-button-content")).getText();
    }
    mouseClickHold() {
        protractor_1.browser.actions()
            .mouseMove(this.getElement().getWebElement())
            .perform();
        protractor_1.browser.actions().mouseDown(this.getElement().getWebElement()).perform();
        protractor_1.browser.sleep(2000);
        protractor_1.browser.actions().mouseUp().perform();
        return;
    }
    ;
}
ButtonAtom.CSS_CLASS = "xui-button";
exports.ButtonAtom = ButtonAtom;
