// @solarwinds/xui/sdk/spec/components/search/search.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
const index_atom_1 = require("../index.atom");
class SearchAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.getValueAttr = () => {
            return super.getElement().element(protractor_1.by.tagName("input")).getAttribute("value");
        };
        this.acceptInput = (input) => {
            return super.getElement().element(protractor_1.by.tagName("input")).sendKeys(input);
        };
        this.areSuggestionsDisplayed = () => {
            return this.getDropdown().isDisplayed();
        };
        this.getSuggestions = () => {
            return this.getDropdown().all(protractor_1.by.css("li"));
        };
        this.getSuggestion = (index) => {
            return this.getDropdown().all(protractor_1.by.css("li")).get(index);
        };
        this.getSuggestionLinks = () => {
            return this.getDropdown().all(protractor_1.by.css("li a"));
        };
        this.getSuggestionLink = (index) => {
            return this.getSuggestionLinks().get(index);
        };
        this.isBusy = () => {
            return this.getElement().element(protractor_1.by.css(".xui-search__input.xui-search--busy")).isPresent();
        };
        this.isActive = () => {
            return this.getElement().element(protractor_1.by.css(".xui-search__input.xui-search--active")).isPresent();
        };
        this.getSpinner = () => {
            return new index_atom_1.SpinnerAtom(this.getElement().element(protractor_1.by.className("xui-search__input-spinner")));
        };
        this.getSearchButton = () => {
            return uzi_1.Atom.findIn(index_atom_1.ButtonAtom, this.getElement().element(protractor_1.by.className("xui-search__button")));
        };
        this.getCancelButton = () => {
            return new index_atom_1.ButtonAtom(this.getElement().element(protractor_1.by.className("xui-search__input-cancel")));
        };
        this.getDropdown = () => {
            return super.getElement().element(protractor_1.by.css("ul.dropdown-menu"));
        };
    }
}
SearchAtom.CSS_CLASS = "xui-search";
exports.SearchAtom = SearchAtom;
