// @solarwinds/xui/ng1/src/components/ifShow/ifShow-controller.ts
import {IScope} from "angular";
import SwUtil from "../../services/swUtil-service";

/**
 * Xui ifShow directive associated controller
 * 
 * @class
 */
export default class IfShowController {
    public static $inject: string[] = [];

    /**
     * Bound indicator whether to show content
     */
    public visible: boolean;

    public showNgIf: boolean;
    public showNgShow: boolean;

    /**
     * Indicator whether content has been shown at least once
     */
    public hasBeenShown: boolean = false;

    public $onChanges = (changes: any): void => {
        if (!changes.visible) {
            return;
        }

        this.onVisibleChanged(changes.visible.currentValue);
    };

    public onVisibleChanged = (visible?: boolean): void => {
        
        if (!this.hasBeenShown) {
            
            // 1st show
            if (visible) {
                this.showNgIf = true;
                this.showNgShow = true;
                this.hasBeenShown = true;
            } else {
                this.showNgIf = false;
                this.showNgShow = false;
            }
        } else {
            // only ng-show is used in upcoming showing/hiding 
            this.showNgShow = visible;
        }
    };
}
