// @solarwinds/xui/sdk/spec/components/sorter/sorter.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const index_atom_1 = require("../index.atom");
class SorterAtom extends index_atom_1.DropdownAtom {
    getItemCount() {
        return super.getElement().all(protractor_1.by.repeater("item in vm.itemsSource")).count();
    }
    getItemText(index) {
        return protractor_1.browser.executeScript("return arguments[0].innerHTML;", super.getElement().all(protractor_1.by.repeater("item in vm.itemsSource")).get(index));
    }
    clickSortToggle() {
        return super.getElement()
            .element(protractor_1.by.className("xui-dropdown__action"))
            .click();
    }
    isAscendingSortActionDisplayed() {
        return protractor_1.element(protractor_1.by.id("sortActionAsc")).isDisplayed();
    }
    isDescendingSortActionDisplayed() {
        return protractor_1.element(protractor_1.by.id("sortActionDesc")).isDisplayed();
    }
}
SorterAtom.CSS_CLASS = "xui-sorter";
exports.SorterAtom = SorterAtom;
