// @solarwinds/xui/ng2/demo/src/components/demo/dragdrop/dragdrop.example.component.ts
import {Component, Inject, OnInit} from "@angular/core";
import {IToastService} from "../../../../../src/services/toast.service";
import * as xui from "xui";
import {IXuiDropEvent} from "../../../../../src/directives/dragdrop/droppable.directive";

@Component({
    selector: "dragdrop-example",
    template: <string>require("./dragdrop.example.component.html")
})
export class DragdropExampleComponent implements OnInit {
    public draggableString = "this is a string";
    public draggableObj: any;
    public destObject: any;
    public destString: any;
    public draggableExcel = "<table>"+
        "<tr>"+
        "<th>thing1</th>"+
        "<th>thing2</th>"+
        "</tr>"+
        "<tr>"+
        "<td>jeff</td>"+
        "<td>4</td>"+
        "</tr>"+
        "<tr>"+
        "<td>john</td>"+
        "<td>8</td>"+
        "</tr>"+
        "</table>";

    constructor(@Inject(xui.ToastService) private toastService: IToastService) {

    }
    ngOnInit() {
        this.draggableObj = {
            imma: "little",
            teapot: "short and stout"
        };
    }

    public onDragStart (event: DragEvent) {
        this.toastService.info("Drag start");
    };

    public onDragEnd (event: DragEvent) {
        this.toastService.info("Drag end");
    };

    public onDragOver (event: DragEvent) {
        this.toastService.info("Drag over", null, {preventDuplicates: true});
    };

    public onDragEnter (event: DragEvent) {
        this.toastService.info("Drag enter");
    };

    public onDragLeave (event: DragEvent) {
        this.toastService.info("Drag leave");
    };

    public onDropObject(dropEvent: IXuiDropEvent) {
        this.toastService.info("Drop!");
        this.destObject = dropEvent.payload;
    };

    public onDropString(dropEvent: IXuiDropEvent) {
        this.destString = JSON.stringify(dropEvent.payload);

    }

    public onDropElement(dropEvent: IXuiDropEvent) {
        console.log(dropEvent);
    }

    public onButtonClick(event: Event) {
        this.toastService.error("Error");
    }
}
