// @solarwinds/xui/ng2/demo/src/components/demo/menu/menu.example.component.ts
import { Component } from "@angular/core";

@Component({
    selector: "menu-example",
    template: <string>require("./menu.example.component.html")
})
export class MenuExampleComponent {

    public actionClicks = 0;

    public linkClicks = 0;

    public disabled = false;

    public checked = false;

    public size = "";

    public icon = "";

    public toggleDisabled(): void {
        this.disabled = !this.disabled;
    }

    public handleActionClick(): void {
        this.actionClicks++;
    }

    public handleChange(event: any): void {
        this.checked = event.target.checked;
    }

    public handleLinkClick(): void {
        this.linkClicks++;
    }

    public handleSizeChange(event: any) {
        this.size = event.target.value;
    }

    public handleIconChange(event: any) {
        this.icon = event.target.value;
    }

    public itemsSource = [
        {itemType: "header", title: "Links:"},
        {itemType: "link", title: "Link", url: "#button"},
        {itemType: "link", title: "Disabled link", disabled: true},
        {itemType: "header", title: "Actions:"},
        {itemType: "action", title: "Action", action: this.actionDone},
        {itemType: "action", title: "Disabled action", disabled: true},
        {itemType: "header", title: "Options:"},
        {itemType: "option", title: "Option", checked: true},
        {itemType: "option", title: "Disabled option", disabled: true},
        {itemType: "header", title: "Hide & Remove"},
        {itemType: "link", title: "Hidden", url: "#button", hidden: true},
        {itemType: "link", title: "Removed", url: "#button", removed: true},
        {itemType: "header", title: "Icons"},
        {itemType: "link", title: "Link with icon", url: "#button", icon: "state_running"}
    ];

    public code = this.toJSON(this.itemsSource);

    private toJSON(obj: any) {
        return JSON.stringify(obj, null, 3);
    }

    actionDone(): void {
        console.log("Action done!");
    }

}
