// @solarwinds/xui/ng1/sdk/spec/components/watchCounter/watchCounter.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class WatchCounterAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.isDisplayed = () => {
            return this.getElement().isDisplayed();
        };
        this.mouseHover = () => {
            return protractor_1.browser.actions()
                .mouseMove(this.getElement().getWebElement())
                .perform();
        };
    }
    static getBadge() {
        return new WatchCounterAtom(protractor_1.element(protractor_1.by.className("xui-watch-counter__highlight")));
    }
    getTextColor() {
        return super.getElement().getCssValue("color");
    }
    getBackgroundColor() {
        return super.getElement().getCssValue("background-color");
    }
    getBorderStyle() {
        return super.getElement().getCssValue("outline");
    }
    getTextNum() {
        return super.getElement().getText();
    }
}
WatchCounterAtom.CSS_CLASS = "xui-watch-counter";
exports.WatchCounterAtom = WatchCounterAtom;
