// @solarwinds/xui/ng1/sdk/spec/components/message/message.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class MessageAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.allowDismiss = () => {
            return this.hasClass("xui-message-allow-dismiss");
        };
        this.getContent = () => {
            const element = super.getElement().element(protractor_1.by.className("xui-message-content"));
            return element.getText();
        };
        this.getDismissButton = () => {
            return super.getElement().element(protractor_1.by.className("xui-message-dismiss-button"));
        };
        this.dismissMessage = () => {
            if (this.allowDismiss()) {
                this.getDismissButton().click();
            }
        };
        this.isVisible = () => {
            return super.getElement().isPresent();
        };
        this.isDismissable = () => {
            return this.getDismissButton().isDisplayed();
        };
    }
}
MessageAtom.CSS_CLASS = "xui-message";
exports.MessageAtom = MessageAtom;
