// @solarwinds/xui/ng1/spec/components/grid/grid.atom.ts
import {by, ElementFinder, promise} from "@solarwinds/uzi/protractor";
import {Atom} from "@solarwinds/uzi";
import {ListViewAtom, PagerAtom, SelectorAtom, SorterAtom, ToolbarAtom} from "../index.atom";
import SpecHelpers from "../specHelpers";

export class GridAtom extends Atom {
    public static CSS_CLASS = "xui-grid";

    public getListView(): ListViewAtom {
        return Atom.findIn(ListViewAtom, this.getElement());
    }

    public getPager(): PagerAtom {
        return Atom.findIn(PagerAtom, this.getElement());
    }

    public getSorter(): SorterAtom {
        return Atom.findIn(SorterAtom, this.getElement());
    }

    public getSelector(): SelectorAtom {
        return Atom.findIn(SelectorAtom, this.getElement());
    }

    public getToolbar(): ToolbarAtom {
        return Atom.findIn(ToolbarAtom, this.getElement());
    }

    public getSelectionTitleElement(): ElementFinder {
        return super.getElement().element(by.className("xui-grid__selection-title"));
    }

    public clickSearchButton(): promise.Promise<void> {
        let searchButton = super.getElement().element(by.className("xui-search__button"));
        return searchButton.click();
    }

    public clickSearchClearButton(): promise.Promise<void> {
        let clearSearchButton = super.getElement().element(by.className("xui-icon-close"));
        return clearSearchButton.click();
    }

    /**
     * Clicks the Clear Search button if present otherwise does nothing.
     */
    public clickSearchClearButtonIfPresent(): promise.Promise<void> {
        return SpecHelpers.wrapPromise<void>((async () => {
            try {
                let clearSearchButton = this.getElement().element(by.className("xui-icon-close"));
                if (await clearSearchButton.isPresent() && await clearSearchButton.isDisplayed()) {
                    return await clearSearchButton.click();
                }
            } catch (e) {
                console.log("GridAtom.clickSearchClearButtonIfPresent", e);
            }
        })());
    }

    public writeIntoSearch(value: string): promise.Promise<void> {
        let inputField = super.getElement().element(by.className("xui-search__input-control"));
        return inputField.sendKeys(value);
    }
}
