// @solarwinds/xui/ng1/sdk/spec/components/grid/grid.atom.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
const index_atom_1 = require("../index.atom");
const specHelpers_1 = require("../specHelpers");
class GridAtom extends uzi_1.Atom {
    getListView() {
        return uzi_1.Atom.findIn(index_atom_1.ListViewAtom, this.getElement());
    }
    getPager() {
        return uzi_1.Atom.findIn(index_atom_1.PagerAtom, this.getElement());
    }
    getSorter() {
        return uzi_1.Atom.findIn(index_atom_1.SorterAtom, this.getElement());
    }
    getSelector() {
        return uzi_1.Atom.findIn(index_atom_1.SelectorAtom, this.getElement());
    }
    getToolbar() {
        return uzi_1.Atom.findIn(index_atom_1.ToolbarAtom, this.getElement());
    }
    getSelectionTitleElement() {
        return super.getElement().element(protractor_1.by.className("xui-grid__selection-title"));
    }
    clickSearchButton() {
        let searchButton = super.getElement().element(protractor_1.by.className("xui-search__button"));
        return searchButton.click();
    }
    clickSearchClearButton() {
        let clearSearchButton = super.getElement().element(protractor_1.by.className("xui-icon-close"));
        return clearSearchButton.click();
    }
    /**
     * Clicks the Clear Search button if present otherwise does nothing.
     */
    clickSearchClearButtonIfPresent() {
        return specHelpers_1.default.wrapPromise((() => __awaiter(this, void 0, void 0, function* () {
            try {
                let clearSearchButton = this.getElement().element(protractor_1.by.className("xui-icon-close"));
                if ((yield clearSearchButton.isPresent()) && (yield clearSearchButton.isDisplayed())) {
                    return yield clearSearchButton.click();
                }
            }
            catch (e) {
                console.log("GridAtom.clickSearchClearButtonIfPresent", e);
            }
        }))());
    }
    writeIntoSearch(value) {
        let inputField = super.getElement().element(protractor_1.by.className("xui-search__input-control"));
        return inputField.sendKeys(value);
    }
}
GridAtom.CSS_CLASS = "xui-grid";
exports.GridAtom = GridAtom;
