// @solarwinds/xui/ng2/src/components/sidebar/sidebar.component.ts
import {Component, Input, SimpleChanges, OnChanges} from "@angular/core";

@Component({
    selector: "xui-sidebar",
    template: require<string>("./sidebar.component.html")
})

/**
 * Sliding block with generic content at the right/left side of the page, requires
 * <strong>xui-sidebar-container</strong> as parent component.
 * <example-url>./../../../demo/index.html#/sidebar</example-url>
 */
export class SidebarComponent {

    /** 
     * Show or hide sidebar
     */
    @Input() display: boolean;

    /** 
     * Display sidebar on the right or left position. Possible values are: "right", "left"
     */
    @Input() position: string;

    /** 
     * Show sidebar over content if shrink = false, show sidebar near content if shrink = true
     */
    @Input() shrink: boolean;

    /**
     * Class that is used for sidebar toggling/sliding
     */
    public slideClass: string;

    /**
     * Class that is used for sidebar showing over content or near it
     */
    public shrinkClass: string;

    /**
     * Component initialization
     */
    ngOnInit(): void { 
        this.syncState();
    }

    ngOnChanges(changes: SimpleChanges) {
        if (changes["display"]) {
            this.syncState();
        }
    }
    /**
    * Set proper css classes to sidebar component according to binding values
    */
    public syncState = () => {
        this.shrinkClass = this.shrink ? "xui-sidebar--shrink" : "";

        this.slideClass = "xui-sidebar-" + (this.shrink ? "-shrink-" : "");

        if (this.display) {
            this.slideClass += "slidein-";
        } else {
            this.slideClass += "slideout-";
        }

        if (this.position === "left") {
            this.slideClass += "left";
        } else {
            this.slideClass += "right";
        }
    };
}
