// @solarwinds/xui/ng1/sdk/spec/components/popover/popover.atom.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
const index_atom_1 = require("../index.atom");
class PopoverAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.moveMouseToTarget = () => protractor_1.browser.actions()
            .mouseMove(this.getElement().getWebElement())
            .perform();
        this.moveMouseAndClickTarget = () => protractor_1.browser.actions()
            .mouseMove(this.getElement().getWebElement())
            .click()
            .perform();
        this.togglePopover = () => this.isPopoverDisplayed().then((wasDisplayed) => {
            this.moveMouseAndClickTarget();
            return protractor_1.browser.wait(() => this.isPopoverDisplayed().then((isDisplayed) => wasDisplayed !== isDisplayed), 10000);
        });
        this.getPopover = () => this.getPopoverUid().then((id) => protractor_1.browser.element(protractor_1.by.xpath("//*[@popover-uid='" + id + "' and contains(@class, 'xui-popover popover')]")));
        this.isPopoverDisplayed = () => this.getPopover().then(popover => popover.isDisplayed());
        this.getTitle = () => this.getPopover().then(popoverElement => popoverElement.element(protractor_1.by.css(".xui-popover__title")));
        this.getTitleText = () => this.getTitle().then(titleElement => titleElement.getText());
        this.hasStatus = (status) => __awaiter(this, void 0, void 0, function* () {
            const title = yield this.getTitle();
            return yield uzi_1.Atom.hasClass(title, "xui-popover__title--" + status);
        });
        this.hasNoStatus = () => __awaiter(this, void 0, void 0, function* () {
            const title = yield this.getTitle();
            return yield uzi_1.Atom.hasClass(title, "xui-popover__title--");
        });
        this.getStatusIconName = () => this.getPopover().then(popover => index_atom_1.IconAtom.findIn(index_atom_1.IconAtom, popover).name());
        this.isDisplayedRight = () => __awaiter(this, void 0, void 0, function* () {
            const popover = yield this.getPopover();
            return yield uzi_1.Atom.hasClass(popover, "right");
        });
        this.isDisplayedLeft = () => __awaiter(this, void 0, void 0, function* () {
            const popover = yield this.getPopover();
            return yield uzi_1.Atom.hasClass(popover, "left");
        });
        this.clickTarget = () => this.getElement().click();
        this.getPopoverUid = () => this.getElement().getAttribute("popover-uid");
    }
}
PopoverAtom.CSS_CLASS = "xui-popover--click";
exports.PopoverAtom = PopoverAtom;
