// @solarwinds/xui/sdk/spec/components/empty/empty.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class EmptyAtom extends uzi_1.Atom {
    getTitle() {
        return this.getEmptyElement("title").getText();
    }
    getDescription() {
        return this.getEmptyElement("description").getText();
    }
    getImage() {
        return this.getEmptyElement("image").getAttribute("image");
    }
    ;
    isDefaultTemplateRendered() {
        return super.getElement().isPresent();
    }
    isCustomTemplateRendered() {
        return protractor_1.browser.element(protractor_1.by.className("sample-custom-template")).isPresent();
    }
    isCompactModeEnabled() {
        return super.hasClass("xui-empty-compact-mode");
    }
    hasTitle() {
        return this.getEmptyElement("title").isPresent();
    }
    getCustomTemplateBoundValueText() {
        return protractor_1.browser.element(protractor_1.by.className("sample-custom-template_bound-value")).getText();
    }
    getEmptyElement(property) {
        return super.getElement()
            .element(protractor_1.by.className(`${EmptyAtom.CSS_CLASS}_${property}`));
    }
}
EmptyAtom.CSS_CLASS = "xui-empty";
exports.EmptyAtom = EmptyAtom;
