// @solarwinds/xui/ng1/spec/components/chiclets/chiclets.atom.ts
import {by, promise} from "@solarwinds/uzi/protractor";
import {WebElementPromise} from "@solarwinds/uzi/webdriver";
import {Atom} from "@solarwinds/uzi";

export class ChicletsAtom extends Atom {
    public static CSS_CLASS = "xui-chiclets";

    public isVisible(): promise.Promise<boolean> {
        return super.getElement().isDisplayed();
    }

    public getClearAllLinkElement(): WebElementPromise {
        return super.getElement().element(by.className("xui-chiclets__clear")).getWebElement();
    }

    public clearAll() {
        return super.getElement().element(by.className("xui-chiclets__clear")).click();
    }

    public getChicletValue(index: number) {
        return super.getElement().all(by.className("xui-chiclets__value-name")).get(index).getText();
    }

    public getChicletsCount() {
        return super.getElement().all(by.css(".xui-chiclets__value")).count();
    }

    public getChicletsNames() {
        return super.getElement().all(by.css(".xui-chiclets__value")).map(el => el.$(".xui-chiclets__value-name")
            .getText());
    }

    public getChicletsGroupCount() {
        return super.getElement().all(by.css(".xui-chiclets__group-name")).count();
    }

    public getChicletsGroupNames() {
        return super.getElement().all(by.css(".xui-chiclets__group-name")).map(el => el.getText());
    }

    public getChicletsQuantityFromLabel() {
        return super.getElement().element(by.css(".xui-chiclets__count")).getText().then(t => +t.slice(1, -1));
    }

    public removeItem(index: number) {
        return super.getElement().all(by.className("xui-chiclets__value-remove")).get(index).click();
    }

}
