// @solarwinds/xui/ng2/src/components/page-content/page-content.component.ts
import {
    Component,
    Inject,
    Input,
    OnChanges,
    SimpleChanges
} from "@angular/core";
import { DOCUMENT } from "@angular/common";

/**
 * <example-url>./../../../demo/index.html#/page-content</example-url>  
 * 
 * Component that provides the basic layout of a page, including breadcrumbs, transcluded header, 
 * body and footer. In essense, component logic is applying some css styles to parent html, that 
 * styling to siblings consquently. Additionally, it provides default template for header if no
 * custom one is provided.
 *  
 * Inner header and footer elements for transclusion should be marked with [xui-page-content\_\_header]
 * and [xui-page-content\_\_footer] attributes accordingly. 
 * The rest of transcluded elements will be transcluded as a body.
 * 
 * If no header provided for transclusion inside this component, default inner template is used.
 * The styling can be altered by applying a class to the container element. Currently supported 
 * class is 'xui-page-content--no-padding'.
 */
@Component({
    selector: "xui-page-content",
    template: require<string>("./page-content.component.html"),
})
export class PageContentComponent implements OnChanges {
    /**
     * Determines default  'pageLayout' prop that is being mapped according to 'PAGE_LAYOUT_MAP'
     */
    public static readonly DEFAULT_PAGELAYOUT: PageContentLayouts = "fullWidth";
    /**
     * Determines how input 'pageLayout' prop is mapped to css clasas
     */
    public static readonly PAGE_LAYOUT_MAP: { [key: string]: string } = {
        "fullWidth": "xui-page-content--full-width",
        "fill": "xui-page-content--fill",
        "form": "xui-page-content--form"
    };

    /**
     * Parameter to control if busy indicator should be displayed
     */
    @Input() isBusy: boolean;

    /**
     * Message displayed along with the busy indicator
     */
    @Input() busyMessage: string;

    /**
     * Title of default header (used if no element with 'xui-page-content__header' attr is 
     * provided for transclusion)
     */
    @Input() pageTitle: string;

    /**
     * link-icon icon displayed to the left of the header link (used if no element with 
     * 'xui-page-content__header' attr is provided for transclusion)
     */
    @Input() linkIcon: string;

    /**
     * link provided in the top-right corner of the default header
     */
    @Input() linkUrl: string;

    /**
     * Link text for defauld header
     */
    @Input() linkText: string;

    /**
     * link-target attribute for the link in default header: '_blank' for new tab, or '_self' 
     * to force navigation
     */
    @Input() linkTarget: string;

    /**
     * Determines how the page is displayed. Options are [form, fullWidth (default), fill].
     */
    @Input() pageLayout: PageContentLayouts;

    constructor(@Inject(DOCUMENT) private document: Document) {
        this.pageLayout = PageContentComponent.DEFAULT_PAGELAYOUT;
    }

    /**
     * reflects whether link is shown or not in default header
     */
    public isLinkShown = () => {
        return !_.isEmpty(this.linkText);
    };

    /**
     * reflects whether link icon is shown or not in default header
     */
    public isIconShown = () => {
        return !_.isEmpty(this.linkIcon);
    };
    
    public ngOnChanges(changes: SimpleChanges) {
        if (changes["pageLayout"]) {
            this.adjustLayout();
        };
    }

    private adjustLayout = () => {
        const $htmlEl: HTMLHtmlElement = this.document.getElementsByTagName("html")[0];
        Object.keys(PageContentComponent.PAGE_LAYOUT_MAP).forEach((key: string) => {
            let className: string = PageContentComponent.PAGE_LAYOUT_MAP[key];
            $htmlEl.classList.remove(className);
        });
        $htmlEl.classList.add(PageContentComponent.PAGE_LAYOUT_MAP[this.pageLayout]);
    }
}

export type PageContentLayouts = "fullWidth" | "fill" | "form";
