// @solarwinds/xui/ng1/sdk/spec/components/switch/switch.atom.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class SwitchAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.container = () => {
            return this.getElement().element(protractor_1.by.className("xui-switch__container"));
        };
        this.caption = () => {
            return this.getElement().element(protractor_1.by.className("xui-switch__label"));
        };
        this.slider = () => {
            return this.getElement().element(protractor_1.by.className("xui-switch__bar"));
        };
        this.isOn = () => {
            return this.hasClass(SwitchAtom.ON_CSS);
        };
        this.isEnabled = () => __awaiter(this, void 0, void 0, function* () {
            const attValue = yield this.getElement().getAttribute("disabled");
            return attValue !== "true";
        });
    }
}
SwitchAtom.ON_CSS = "xui-switched";
SwitchAtom.CSS_CLASS = "xui-switch";
exports.SwitchAtom = SwitchAtom;
