// @solarwinds/xui/ng2/src/components/toast/toast-ref.ts
import { Subject } from "rxjs/Subject";
import { ToastContainerService } from "../../services/toast-container.service";
import { Observable } from "rxjs/Observable";

/**
 * Reference to a toast opened via the Toastr service.
 * Used to provide toast events hooks
 */
export class ToastRef<T> {
    /** The instance of component opened into the toast. */
    componentInstance: T;
    /** Subject for notifying the user that the toast has finished closing. */
    private afterClosedSubject: Subject<any> = new Subject();
    private activateSubject: Subject<any> = new Subject();
    private manualCloseSubject: Subject<any> = new Subject();

    constructor(private toastContainerService: ToastContainerService) {}

    public manualClose() {
        this.manualCloseSubject.next();
        this.manualCloseSubject.complete();
    }

    public manualClosed(): Observable<any> {
        return this.manualCloseSubject.asObservable();
    }

    /**
     * Close the toast
     */
    public close(): void {
        this.toastContainerService.detachToast(this.componentInstance);
        this.afterClosedSubject.next();
        this.afterClosedSubject.complete();
    }

    /** Gets an observable that is notified when the toast is finished closing. */
    public afterClosed(): Observable<any> {
        return this.afterClosedSubject.asObservable();
    }

    public isInactive() {
        return this.activateSubject.isStopped;
    }

    public activate() {
        this.activateSubject.next();
        this.activateSubject.complete();
    }

    /** Gets an observable that is notified when the toast has started opening. */
    public afterActivate(): Observable<any> {
        return this.activateSubject.asObservable();
    }
}
