// @solarwinds/xui/ng2/spec/components/checkbox/checkbox.atom.ts
import {by, ElementFinder, promise} from "@solarwinds/uzi/protractor";
import {Atom} from "@solarwinds/uzi";

export class CheckboxAtom extends Atom {
    public static CSS_CLASS = "xui-checkbox";

    constructor(element: ElementFinder) {
        super(element);
    }

    public getInputElement(): ElementFinder {
        return super.getElement().element(by.className("xui-checkbox__input"));
    };

    public getLabel(): ElementFinder {
        return super.getElement().element(by.className("xui-checkbox__label"));
    };

    public getContent(): promise.Promise<string> {
        const element = super.getElement().element(by.className("xui-checkbox__transclude"));
        return element.getText();
    };

    public getHelpHintText(): promise.Promise<string> {
        const element = super.getElement().element(by.className("xui-input-help"));
        return element.getText();
    };

    public getValidationText(): promise.Promise<string> {
        const element = super.getElement().element(by.className("xui-validation"));
        return element.getText();
    };

    public isInvalid(): promise.Promise<boolean> {
        return this.getInputElement().getAttribute("aria-invalid").then((value) => {
            return value === "true";
        });
    };

    public isIndeterminate(): promise.Promise<boolean> {
        return this.getInputElement().getAttribute("indeterminate").then((value) => {
            return value === "true";
        });
    };

    public isRequired(): promise.Promise<boolean> {
        return this.getInputElement().getAttribute("required").then((value) => {
            return value === "true";
        });
    };

    public isChecked(): promise.Promise<boolean> {
        const cbInput = this.getInputElement();
        return cbInput.getAttribute("checked").then((value) => {
            return value === "true";
        });
    };

    public isDisabled(): promise.Promise<boolean> {
        return this.getInputElement().isEnabled().then((value) => {
            return !value;
        });
    };

    public click(): promise.Promise<void> {
        return this.getMark().click();
    };

    private getMark(): ElementFinder {
        return this.getElement().element(by.className("xui-checkbox__mark"));
    };
}
