// @solarwinds/xui/ng2/src/constants/empty.constants.ts
import { InjectionToken } from "@angular/core";

export interface IEmptyData {
    preset?: EmptyPresetItemName;
    compactMode?: boolean;
    image?: string;
    title?: string;
    description?: string;
    template?: string;
}

export type EmptyPresetItemName = "dataNotAvailable"
    | "serverUnreachable"
    | "dataNotApplicable"
    | "emptyDataset";

export interface IEmptyPreset {
    [key: string]: IEmptyData;
}

export const EMPTY_PRESET = {
    dataNotAvailable: {
        compactMode: true,
        image: "no-data-to-show",
        title: "",
        description: "Data is not available",
        template: null
    } as IEmptyData,
    serverUnreachable: {
        compactMode: true,
        image: "no-data-to-show",
        title: "",
        description: "Server unreachable",
        template: null
    } as IEmptyData,
    dataNotApplicable: {
        compactMode: true,
        image: "no-data-to-show",
        title: "",
        description: "Data is not applicable",
        template: null
    } as IEmptyData,
    emptyDataset: {
        compactMode: true,
        image: "no-data-to-show",
        title: "",
        description: "No data to show",
        template: null
    } as IEmptyData
};

export const EmptyPresetToken = new InjectionToken<IEmptyPreset>("EMPTY_PRESET");
