// @solarwinds/xui/ng1/src/components/listitem/listitem-directive.ts
import Inject from "../../decorators/di";

/**
 * @ngdoc directive
 * @name xui.directive:xuiListitem
 * @restrict E
 *
 * @description
 * XUI wrapper for listitem on lists.
 *
 * @example
 *    <example module="xui">
 *        <file src="src/components/listitem/docs/listitem-examples.html" name="index.html"></file>
 *    </example>
 */
export default class Listitem implements ng.IDirective {
    constructor(@Inject("swUtil") private swUtil: any) {
    }

    public restrict = "E";
    public templateUrl = "xui/components/listitem/listitem-directive.html";
    public replace = true;
    public transclude = true;
    public scope = {};
    public controller = "xuiListitemController";
    public controllerAs = "vmListItem";
    public bindToController = {
        name: "@?",
        role: "@?",
        isDisabled: "=?",
        isSelected: "<?",
        isClickable: "<?",
        paddingSize: "@?",
        nowrap: "=?"
    };

    public compile = (element: ng.IAugmentedJQuery, attrs: ng.IAttributes) => {
        if ("isDisabled" in attrs) {
            element.attr("ng-disabled", "ctrl.isDisabled");
        }

        if (!attrs["role"]) {
            element.attr("role", "listitem");
        }
        return {post: this.postLink};
    };

    private postLink = (scope: ng.IScope,
                        element: ng.IAugmentedJQuery,
                        attrs: ng.IAttributes) => {
        this.swUtil.initComponent(attrs, "listitem");
    };
}
