// @solarwinds/xui/ng2/demo/src/components/demo/sidebar/sidebar.example.component.ts
import { Component } from "@angular/core";

@Component({
    template: require<string>("./sidebar.example.component.html"),
    selector: "xui-sidebar-example",
})

export class SidebarExampleComponent {

    settings = {
        display: false,
        position: "left",
        shrink: false
    };

    settings2 = {
        display: false,
        position: "right",
        shrink: true
    };

    dynamicContent: string = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras at viverra orci. " +
        "Sed elementum lacus eget finibus varius. Fusce rhoncus libero augue. " +
        "Cras cursus purus convallis ex laoreet, a pulvinar neque mollis. " +
        "Maecenas fermentum, turpis non aliquam suscipit, sapien ligula feugiat lorem, a suscipit nulla " +
        "Sed mollis iaculis varius. Suspendisse auctor finibus diam, eget laoreet tellus fermentum nec. " +
        "Fusce iaculis varius lorem at lacinia. Curabitur et cursus eros, at congue lorem. " +
        "Aenean urna ipsum, sollicitudin sed ultrices nec, malesuada eu lacus. " +
        "Sed orci mauris, scelerisque vitae consectetur sit amet, venenatis a mauris. " +
        "Donec consectetur libero eget bibendum consequat. Donec tincidunt molestie viverra.";

    toggleSidebar () {
        this.settings.display = !this.settings.display;
    };

    toggleSidebar2 () {
        this.settings2.display = !this.settings2.display;
    };
}
