// @solarwinds/xui/ng1/sdk/spec/components/codeItem/codeItem.atom.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class CodeItemAtom extends uzi_1.Atom {
    toHex(rgb_color) {
        let rgb = rgb_color.match(/^rgba?[\s+]?\([\s+]?(\d+)[\s+]?,[\s+]?(\d+)[\s+]?,[\s+]?(\d+)[\s+]?/i);
        return (rgb && rgb.length === 4) ? "#" +
            ("0" + parseInt(rgb[1], 10).toString(16)).slice(-2) +
            ("0" + parseInt(rgb[2], 10).toString(16)).slice(-2) +
            ("0" + parseInt(rgb[3], 10).toString(16)).slice(-2) : "";
    }
    getTextByClassName(selector) {
        return super.getElement().element(protractor_1.by.className(selector)).getText();
    }
    getTextFromLineNumber(selector) {
        //Since text from line-numbers can no longer be selected, we need to select content from 'before' pseudo element
        let script = `window.getComputedStyle(document.querySelectorAll(".${selector}")[0], ":before").content`;
        return protractor_1.browser.executeScript(`return ${script}`);
    }
    getBackgroundColor() {
        return super.getElement().getCssValue("background-color");
    }
    getBgColorAndHexIt() {
        const _super = name => super[name];
        return __awaiter(this, void 0, void 0, function* () {
            let color = yield _super("getElement").call(this).getCssValue("background-color");
            return this.toHex(color);
        });
    }
    getTextColor(className) {
        const _super = name => super[name];
        return __awaiter(this, void 0, void 0, function* () {
            return yield _super("getElement").call(this).element(protractor_1.by.className(className)).getCssValue("color");
        });
    }
    getTextColorAndHexIt(className) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.toHex(yield this.getTextColor(className));
        });
    }
    getTranscludedColorAndHexIt() {
        const _super = name => super[name];
        return __awaiter(this, void 0, void 0, function* () {
            return this.toHex(yield _super("getElement").call(this).element(protractor_1.by.tagName("strong")).getCssValue("color"));
        });
    }
    getSelectedText() {
        return this.doSelect()
            .then(() => {
            return protractor_1.browser.executeScript(() => {
                let text = "";
                if (window.getSelection) {
                    text = window.getSelection().toString();
                }
                return text;
            });
        });
    }
    doSelect() {
        return protractor_1.browser.actions()
            .sendKeys(protractor_1.protractor.Key.chord(protractor_1.protractor.Key.CONTROL, "a"))
            .perform();
    }
}
CodeItemAtom.CSS_CLASS = "xui-code-item";
exports.CodeItemAtom = CodeItemAtom;
