// @solarwinds/xui/ng2/src/components/checkbox/checkbox.component.ts
import {
    Component,
    Input,
    Output,
    EventEmitter,
    ViewChild,
    ViewContainerRef,
    ChangeDetectionStrategy
} from "@angular/core";

/**
 * Class that is used for storing of xui-checkbox changed value
 */
export class CheckboxChangeEvent {
    /**
     * Changed xui-checkbox component
     */
    public target: CheckboxComponent;
}

@Component({
    selector: "xui-checkbox",
    template: require<string>("./checkbox.component.html"),
    changeDetection: ChangeDetectionStrategy.OnPush
})

/**
 * <xui-checkbox> provides the same functionality as a native 
 * <input type="checkbox"> enhanced with XUI styling.
 * <example-url>./../../../demo/index.html#/checkbox</example-url>
 */
export class CheckboxComponent {
    /**
     * Sets "name" attribute for inner input element of xui-checkbox
     */
    @Input() public name: string;

    /**
     * Sets "title" attribute for xui-checkbox label
     */
    @Input() public title: string = "";

    /**
     * Sets "value" attribute for inner input element of xui-checkbox
     */
    @Input() public value: any;

    /**
     * Event that is fired when xui-checkbox is toggled
     */
    @Output() public valueChange = new EventEmitter<CheckboxChangeEvent>();

    /**
     * Used for changing xui-checkbox component styles when component is hovered
     */
    @Input() public hovered: boolean;

    /**
     * Sets "checked" attribute for inner input element of xui-checkbox
     */
    @Input() public checked: boolean;

    /**
     * Sets "required" attribute for inner input element of xui-checkbox
     */
    @Input() public required: boolean;

    /**
     * Sets xui-checkbox hint that is displayed below label
     */
    @Input() public hint: string;

    /**
     * Used for disabling of xui-checkbox
     */
    @Input() public disabled: boolean;

    /*
     * Text that is displayed when checkbox should be checked
     */
    @Input() public requireText: string;

    /**
     * Used for setting of xui-checkbox indeterminate state
     */
    @Input() public indeterminate: boolean;

    /**
     * Reference to the inner input checkbox html element 
     */
    @ViewChild("inputViewContainer", { read: ViewContainerRef })
    public inputViewContainer: ViewContainerRef;

    /**
     * Used for changing of css style when xui-checkbox is hovered
     */
    public hoverHandler() {
        this.hovered = !this.hovered;
    }

    /**
     * xui-checkbox check/uncheck state handler
     * @param event Changed xui-checkbox component
     */
    public changeHandler(event: CheckboxChangeEvent) {
        this.valueChange.emit(event);
    }

}
