// @solarwinds/xui/sdk/spec/components/toast/toast.atom.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
const specHelpers_1 = require("../specHelpers");
class ToastAtom extends uzi_1.Atom {
    static get(index = 0) {
        return new ToastAtom(protractor_1.element.all(protractor_1.by.css(`.${ToastAtom.CSS_CLASS}`)).get(index));
    }
    static close(waitForNotDisplayed = false) {
        return specHelpers_1.default.wrapPromise((() => __awaiter(this, void 0, void 0, function* () {
            try {
                const toastAtom = ToastAtom.get();
                if (yield toastAtom.getElement().isDisplayed()) {
                    const closed = yield toastAtom.close(waitForNotDisplayed);
                    return closed;
                }
            }
            catch (e) {
                console.log("static ToastAtom.close", e);
            }
            return false;
        }))());
    }
    getTitle() {
        return this.getElement().element(protractor_1.by.css(".toast-title")).getText();
    }
    getMessage() {
        return this.getMessageElement().getText();
    }
    getMessageElement() {
        return this.getElement().element(protractor_1.by.css(".toast-message"));
    }
    isToastType(type) {
        return this.getElement().getAttribute("class")
            .then((className) => className.indexOf(`xui-toast--${type}`) > -1);
    }
    close(waitForNotDisplayed = false) {
        return specHelpers_1.default.wrapPromise((() => __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.getElement().element(protractor_1.by.className("toast-close-button")).click();
                if (!waitForNotDisplayed) {
                    return true;
                }
                const closed = yield this.waitUntilNotDisplayed();
                return closed;
            }
            catch (e) {
                console.log("ToastAtom.close", e);
            }
            return false;
        }))());
    }
    isSuccessType() {
        return this.isToastType("success");
    }
    isWarningType() {
        return this.isToastType("warning");
    }
    isInfoType() {
        return this.isToastType("info");
    }
    isErrorType() {
        return this.isToastType("error");
    }
    waitUntilDisplayed() {
        return specHelpers_1.default.wrapPromise((() => __awaiter(this, void 0, void 0, function* () {
            try {
                return yield protractor_1.browser.wait(() => this.getMessageElement().isDisplayed(), ToastAtom.toastTimeout);
            }
            catch (e) {
                console.log("Toast.waitUntilDisplayed", e);
            }
            return false;
        }))());
    }
    waitUntilNotDisplayed() {
        return specHelpers_1.default.wrapPromise((() => __awaiter(this, void 0, void 0, function* () {
            try {
                return yield protractor_1.browser.wait(() => specHelpers_1.default.isNotPresentOrNotDisplayed(this), ToastAtom.toastTimeout);
            }
            catch (e) {
                console.log("Toast.waitUntilNotDisplayed", e);
            }
            return false;
        }))());
    }
}
ToastAtom.CSS_CLASS = "xui-toast";
ToastAtom.toastTimeout = 7000;
exports.ToastAtom = ToastAtom;
