// @solarwinds/xui/ng1/sdk/spec/components/chiclets/chiclets.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class ChicletsAtom extends uzi_1.Atom {
    isVisible() {
        return super.getElement().isDisplayed();
    }
    getClearAllLinkElement() {
        return super.getElement().element(protractor_1.by.className("xui-chiclets__clear")).getWebElement();
    }
    clearAll() {
        return super.getElement().element(protractor_1.by.className("xui-chiclets__clear")).click();
    }
    getChicletValue(index) {
        return super.getElement().all(protractor_1.by.className("xui-chiclets__value-name")).get(index).getText();
    }
    getChicletsCount() {
        return super.getElement().all(protractor_1.by.css(".xui-chiclets__value")).count();
    }
    getChicletsNames() {
        return super.getElement().all(protractor_1.by.css(".xui-chiclets__value")).map(el => el.$(".xui-chiclets__value-name")
            .getText());
    }
    getChicletsGroupCount() {
        return super.getElement().all(protractor_1.by.css(".xui-chiclets__group-name")).count();
    }
    getChicletsGroupNames() {
        return super.getElement().all(protractor_1.by.css(".xui-chiclets__group-name")).map(el => el.getText());
    }
    getChicletsQuantityFromLabel() {
        return super.getElement().element(protractor_1.by.css(".xui-chiclets__count")).getText().then(t => +t.slice(1, -1));
    }
    removeItem(index) {
        return super.getElement().all(protractor_1.by.className("xui-chiclets__value-remove")).get(index).click();
    }
}
ChicletsAtom.CSS_CLASS = "xui-chiclets";
exports.ChicletsAtom = ChicletsAtom;
