// @solarwinds/xui/ng2/src/components/treeview/treeview-item.component.ts
import { Component,
         Input,
         Output,
         EventEmitter,
         ElementRef } from "@angular/core";

import {IVirtualNode} from "./treeview.service";
import {TreeviewItemMouseClickEvent} from "./treeview.component";

@Component({
    selector: "xui-treeview-item",
    template: require<string>("./treeview-item.component.html")
})
export class TreeviewItemComponent {
    @Input() public itemWithChildrenTemplate: ElementRef;
    @Input() public itemWithoutChildrenTemplate: ElementRef;
    @Input() public node: IVirtualNode;
    @Output() public mouseClick = new EventEmitter<TreeviewItemMouseClickEvent>();

    public onMouseClick($event: MouseEvent) {
        this.mouseClick.emit(<TreeviewItemMouseClickEvent>{event: $event, sourceNode: this.node});
    }

    public onMouseClickBubbling($event: TreeviewItemMouseClickEvent) {
        this.mouseClick.emit($event);
    }

    //TODO: to get rid of twice firing it from xui-expander
    //TODO: consider firing event for node selection when click on expander icon 
    public onCollapsedChange($event: boolean) {
        this.node.collapsed = !$event;
    }
}
