// @solarwinds/xui/sdk/spec/components/fileUpload/fileUpload.atom.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class FileUploadAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.getInput = () => {
            return super.getElement().element(protractor_1.by.css("input[type='file']"));
        };
        this.getBrowseButton = () => {
            return super.getElement().element(protractor_1.by.className("xui-file-upload__browse"));
        };
        this.getChangeButton = () => {
            return super.getElement().element(protractor_1.by.className("xui-file-upload__change"));
        };
        this.getRemoveButton = () => {
            return super.getElement().element(protractor_1.by.className("xui-file-upload__remove"));
        };
        this.getNofileHint = () => {
            return super.getElement().element(protractor_1.by.className("xui-help-hint"));
        };
        this.getFilenameLabel = () => {
            return super.getElement().element(protractor_1.by.className("xui-file-upload__file-name"));
        };
        this.getDisplayedFileName = () => {
            return this.getFilenameLabel().getText();
        };
    }
    selectFile(path) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.getInput().sendKeys(path);
        });
    }
}
FileUploadAtom.CSS_CLASS = "xui-file-upload";
exports.FileUploadAtom = FileUploadAtom;
