// @solarwinds/xui/sdk/spec/components/tabs/tabs.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class TabsAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.tabCount = () => {
            return this.tabs().count();
        };
        this.getTabTitle = (index) => {
            return this.tabs().get(index).element(protractor_1.by.css("a")).getText();
        };
        this.isTabSelected = (index) => {
            return this.tabs().get(index)
                .getAttribute("class")
                .then((css) => {
                return css.search("xui-tab") !== -1 && css.search("active") !== -1;
            });
        };
        this.getContent = (index) => {
            return super.getElement().all(protractor_1.by.css(".xui-tabs__content")).get(index);
        };
        this.getRouterView = () => {
            return super.getElement().element(protractor_1.by.css(".xui-tabs-router-view"));
        };
        this.click = (index) => {
            return this.tabs().get(index).click();
        };
        this.tabs = () => {
            return super.getElement().all(protractor_1.by.css(".nav.nav-pills li"));
        };
    }
    getLeftCaret() {
        return super.getElement().element(protractor_1.by.css(".xui-tabs__header-caret--left"));
    }
    getRightCaret() {
        return super.getElement().element(protractor_1.by.css(".xui-tabs__header-caret--right"));
    }
    getTabsSlidingPanel() {
        return super.getElement().element(protractor_1.by.css(".nav"));
    }
    getTranslateX(elem) {
        //expression for analyzing css transform property value, extracts translateX value
        const transXExp = /^matrix\((?:-?\d+, ){4}-?(\d+), -?\d+\)$/;
        return elem.getCssValue("transform").then((transform) => {
            if (transXExp.test(transform)) {
                return +transXExp.exec(transform)[1];
            }
            else {
                return 0;
            }
        });
    }
}
TabsAtom.CSS_CLASS = "xui-tabs";
exports.TabsAtom = TabsAtom;
