// @solarwinds/xui/ng2/demo/src/components/demo/busy/busy-example.component.ts
import { Component, OnDestroy } from "@angular/core";

@Component({
    selector: "xui-busy-example",
    template: require<string>("./busy-example.component.html")
})
export class BusyExampleComponent implements OnDestroy {

    public busy: boolean;
    public percent: number;
    public message: string;
    public allowCancel: boolean;
    public showProgress: boolean;
    public helpText: string;
    public cancelText: string;

    private interval: any;
    private timeout: any;

    public saveNotes() {
        if (this.interval) {
            return;
        }

        this.busy = true;
        this.percent = 0;
        this.message = "Saving...";
        this.allowCancel = true;
        this.showProgress = true;
        this.helpText = "It won't take long";
        this.cancelText = "cancel";

        this.interval = setInterval(() => {
            if (this.percent < 100) {
                this.percent += 20;
            } else {
                this.stopProgress(true);
            }
        }, 500);
    };

    public onCancel() {
        this.timeout = setTimeout(() => {
            this.busy = false;
        }, 500);

        this.stopProgress(false);
    }

    public ngOnDestroy() {
        clearInterval(this.interval);
        clearTimeout(this.timeout);
    }

    private stopProgress(withDelay: boolean) {
        clearInterval(this.interval);
        this.interval = 0;

        if (withDelay) {
            this.busy = false;
        }

        this.message = "Canceling...";
        this.showProgress = false;
        this.helpText = "";
        this.cancelText = "";
        this.allowCancel = false;
        this.percent = 0;
    }

}
