// @solarwinds/xui/ng2/demo/src/components/demo/empty/empty.example.component.ts
import { Component, Inject } from "@angular/core";

import * as xui from "xui";
import * as _ from "lodash";
import { IEmptyPreset, IEmptyData, EmptyPresetItemName } from "../../../../../src/constants";

@Component({
    selector: "xui-empty-example",
    template: require<string>("./empty.example.component.html")
})
export class EmptyExampleComponent {
    public emptyPresetItemNames: Array<EmptyPresetItemName>;
    public customMode = {
        image: "no-search-results",
        title: "Nothing found",
        description: "Try searching for something else"
    } as IEmptyData;
    public customCompactMode = {
        compactMode: true,
        image: "no-alerts",
        title: "",
        description: "There are no active alerts. Great job."
    } as IEmptyData;

    private customTemplateViewModel = {
        image: "http://oriondemo.solarwinds.com/ui/modules/platform/images/404-Robot.svg",
        searchTerm: "Node134"
    };
    private customTemplate =
        `<div class="empty-container  sample-custom-template"> 
            <img src="${this.customTemplateViewModel.image}"
                 alt="No items in the list" class="empty-pic"/> 
            <p>
                 No result found for 
                 <span class="sample-custom-template_bound-value">${this.customTemplateViewModel.searchTerm}</span>
            </p> 
        </div>`;

    public customTemplateData = {
        template: this.customTemplate
    } as IEmptyData;

    constructor(@Inject(xui.EmptyPresetToken) private emptyPreset: IEmptyPreset) {
        this.emptyPresetItemNames = _.keys(emptyPreset) as Array<EmptyPresetItemName>;
    }

    public getEmptyPresetItem(item: EmptyPresetItemName): IEmptyData {
        return _.get(this.emptyPreset, item as string);
    }
}
