// @solarwinds/xui/ng1/src/components/menu/menu-link-directive.ts
import { IMenuScope } from "./menu-directive";
import MenuItem from "./menu-item";

/**
 * @ngdoc directive
 * @name xui.directive:xuiMenuLink
 * @restrict E
 *
 * @description
 * XUI wrapper for menu link. It is used as hypertext link. It can be used only inside xuiMenu directive.
 *
 * @parameters
 * @param {string} url Value to be used as hypertext link url.
 * @param {string} target Value to be used as link target attribute.
 * @param {boolean} is-disabled Expression to determine if menu item should be disabled.
 *
 * @example
 *    <example module="xui">
 *        <file src="src/components/menu/docs/menu-link-examples.html" name="index.html"></file>
 *        <file src="src/components/menu/docs/menu-link-examples.js" name="app.js"></file>
 *    </example>
 */

interface IMenuLinkScope extends IMenuScope {
    isDisabled: boolean;
    handleClick: Function;
}

export default class MenuLink extends MenuItem implements ng.IDirective {
    public restrict = "E";
    public templateUrl = "xui/components/menu/menu-link-directive.html";
    public replace = true;
    public transclude = true;
    public scope = {
        url: "@?",
        target: "@?",
        isDisabled: "=?",
        icon: "@?"
    };

    public link: ng.IDirectiveLinkFn = (scope: IMenuLinkScope,
                                        element: ng.IAugmentedJQuery,
                                        attrs: ng.IAttributes): void => {
        this.itemType = "link";
        this.addIcon(attrs, element, scope);

        scope.handleClick = (event: MouseEvent) => {
            this.disabled = scope.isDisabled;
            this.menuService.handleItemClick(event, this);
        };
    };
}
