// @solarwinds/xui/ng1/src/components/menu/menu-option-directive.ts
import { IMenuScope } from "./menu-directive";
import MenuItem from "./menu-item";

/**
 * @ngdoc directive
 * @name xui.directive:xuiMenuOption
 * @restrict E
 *
 * @description
 * XUI wrapper for menu option. It is used to display selectable option. It can be used only inside xuiMenu directive.
 *
 * @parameters
 * @param {boolean} is-checked Value which indicates if menu item is selected.
 * @param {boolean} is-disabled Expression to determine if menu item should be disabled.
 *
 * @example
 *    <example module="xui">
 *        <file src="src/components/menu/docs/menu-option-examples.html" name="index.html"></file>
 *        <file src="src/components/menu/docs/menu-option-examples.js" name="app.js"></file>
 *    </example>
 */

interface IMenuOptionScope extends IMenuScope {
    handleClick: Function;
}

export default class MenuOption extends MenuItem implements ng.IDirective {

    public restrict = "E";
    public templateUrl = "xui/components/menu/menu-option-directive.html";
    public replace = true;
    public transclude = true;
    public scope = {
        isChecked: "=?",
        isDisabled: "=?"
    };

    public link: ng.IDirectiveLinkFn = (scope: IMenuOptionScope, element: ng.IAugmentedJQuery,
        attrs: ng.IAttributes): void => {
        this.itemType = "option";

        scope.handleClick = (event: MouseEvent) => {
            this.menuService.handleItemClick(event, this);
        };
    };
}
