// @solarwinds/xui/ng1/src/components/menu/menu-directive.ts
import MenuController from "./menu-controller";
import {IMenuController} from "./menu-controller";

/**
 * @ngdoc directive
 * @name xui.directive:xuiMenu
 * @restrict E
 *
 * @description
 * XUI wrapper for menu control. It allows the ability to display a menu with links, actions, and options. Its items
 * can be separated by dividers, and groups of items can have their own header.
 *
 * @parameters
 * @param {string} title Value to be shown as menu button text.
 * @param {string} [size='normal'] Size of menu. Values: 'normal' or 'small'.
 * @param {string} [display-style='tertiary'] Style used for menu button. Values: 'tertiary' or 'secondary'.
 * @param {boolean} is-disabled Expression to determine if menu should be disabled.
 * @param {string} items-source Value to be used as menu items data.
 * @param {string} [menu-align='left'] Alignment used for menu list. Values: 'left' or 'right'.
 * @param {string} icon CSS class for the menu item.
 * @param {boolean=} menu-append-to-body Appends the menu to '$body' instead of the parent element.
 *
 * @example
 *    <example module="xui">
 *        <file src="src/components/menu/docs/menu-examples.html" name="index.html"></file>
 *        <file src="src/components/menu/docs/menu-examples.js" name="app.js"></file>
 *    </example>
 */
export interface IMenuScope extends ng.IScope {
}

export default class Menu implements ng.IDirective {
    public static $inject = ["swUtil"];

    constructor(private swUtil:any) {
    }

    public restrict = "E";
    public templateUrl = "xui/components/menu/menu-directive.html";
    public replace = true;
    public transclude = true;
    public scope = {};
    public controller = MenuController;
    public controllerAs = "vm";
    public bindToController = {
        title: "@?",
        titleIcon: "@?",
        size: "@?",
        displayStyle: "@?",
        icon: "@?",
        menuAlign: "@?",
        itemsSource: "<?",
        isDisabled: "=?",
        menuAppendToBody: "<?"
    };

    public link:ng.IDirectiveLinkFn =
        (scope:IMenuScope, element:ng.IAugmentedJQuery,
         attrs:ng.IAttributes, menuCtrl:IMenuController):void => {
            //name
            this.swUtil.initComponent(attrs, "menu");
        };
}
