// @solarwinds/xui/ng1/demo/src/content.ts
///<reference path="../../src/ref.d.ts"/>

export = [
    //{
    //    id: "accordion",
    //    name: "Accordion",
    //    url: "/accordion",
    //    templateUrl: "src/templates/components/Accordion.html",
    //    category: "Components"
    //},
    {
        id: "actionDrawer",
        name: "Action drawer",
        url: "/action-drawer",
        template: require("../../src/components/actionDrawer/docs/actionDrawer-examples.html"),
        category: "Components"
    },
    {
        id: "button",
        name: "Button",
        url: "/button",
        template: require("../../src/components/button/docs/button-examples.html"),
        category: "Components"
    },
    {
        id: "busy",
        name: "Busy",
        url: "/busy",
        template: require("../../src/components/busy/docs/busy-examples.html"),
        category: "Components"
    },
    {
        id: "checkbox",
        name: "Checkbox",
        url: "/checkbox",
        template: require("../../src/components/checkbox/docs/checkbox-examples.html"),
        category: "Components"
    },
    {
        id: "checkboxGroup",
        name: "Checkbox group",
        url: "/checkboxgroup",
        template: require("../../src/components/checkbox/docs/checkboxGroup-examples.html"),
        category: "Components"
    },
    {
        id: "clipboard",
        name: "Clipboard",
        url: "/clipboard",
        template: require("../../src/components/clipboard/docs/clipboard-examples.html"),
        category: "Components"
    },
    {
        id: "grid",
        name: "Grid",
        url: "/grid",
        template:  require("../../src/components/grid/docs/grid-examples.html"),
        category: "Components"
    },
    {
        id: "panel",
        name: "Panel",
        url: "/panel",
        template: require("../../src/components/panel/docs/panel-examples.html"),
        category: "Components"
    },
    {
        id: "sorter",
        name: "Sorter",
        url: "/sorter",
        template: require("../../src/components/sorter/docs/sorter-examples.html"),
        category: "Components"
    },
    {
        id: "selector",
        name: "Selector",
        url: "/selector",
        template: require("../../src/components/selector/docs/selector-examples.html"),
        category: "Components"
    },
    {
        id: "statusFilter",
        name: "StatusFilter",
        url: "/statusFilter",
        template: require("../../src/components/statusFilter/docs/statusFilter-examples.html"),
        category: "Components"
    },
    {
        id: "switch",
        name: "Switch",
        url: "/switch",
        template: require("../../src/components/switch/docs/switch-examples.html"),
        category: "Components"
    },
    {
        id: "textbox",
        name: "Textbox",
        url: "/textbox",
        template: require("../../src/components/textbox/docs/textbox-examples.html"),
        category: "Components"
    },
    {
        id: "dropdown",
        name: "Dropdown",
        url: "/dropdown",
        template: require("../../src/components/dropdown/docs/dropdown-examples.html"),
        category: "Components"
    },
    {
        id: "ellipsis",
        name: "Ellipsis",
        url: "/Ellipsis",
        template: require("../../src/components/ellipsis/docs/ellipsis-examples.html"),
        category: "Components"
    },
    {
        id: "empty",
        name: "Empty",
        url: "/Empty",
        template: require("../../src/components/empty/docs/empty-examples.html"),
        category: "Components"
    },
    {
        id: "search",
        name: "Search",
        url: "/search",
        template: require("../../src/components/search/docs/search-examples.html"),
        category: "Components"
    },
    {
        id: "dialog",
        name: "Dialog",
        url: "/dialog",
        template: require("../../src/components/dialog/docs/dialog-examples.html"),
        category: "Components"
    },
    {
        id: "popover",
        name: "Popover",
        url: "/popover",
        template: require("../../src/components/popover/docs/popover-examples.html"),
        category: "Components"
    },
    {
        id: "expander",
        name: "Expander",
        url: "/expander",
        template: require("../../src/components/expander/docs/expander-examples.html"),
        category: "Components"
    },
    {
        id: "fileUpload",
        name: "File Upload",
        url: "/fileupload",
        template: require("../../src/components/fileUpload/docs/fileUpload-examples.html"),
        category: "Components"
    },

    {
        id: "divider",
        name: "Divider",
        url: "/divider",
        template: require("../../src/components/divider/docs/divider-examples.html"),
        category: "Components"
    },
    {
        id: "radio",
        name: "Radio",
        url: "/radio",
        template: require("../../src/components/radio/docs/radio-examples.html"),
        category: "Components"
    },
    {
        id: "listview",
        name: "Listview",
        url: "/listview",
        template: require("../../src/components/listview/docs/listview-examples.html"),
        category: "Components"
    },
    {
        id: "datepicker",
        name: "Date picker",
        url: "/datepicker",
        template: require("../../src/components/datePicker/docs/datePicker-examples.html"),
        category: "Components"
    },
    {
        id: "menu",
        name: "Menu",
        url: "/menu",
        template: require("../../src/components/menu/docs/menu-examples.html"),
        category: "Components"
    },
    {
        id: "message",
        name: "Message",
        url: "/message",
        template: require("../../src/components/message/docs/message-examples.html"),
        category: "Components"
    },
    {
        id: "chiclets",
        name: "Chiclets",
        url: "/chiclets",
        template: require("../../src/components/chiclets/docs/chiclets-examples.html"),
        category: "Components"
    },
    {
        id: "progress",
        name: "Progress",
        url: "/progress",
        template: require("../../src/components/progress/docs/progress-examples.html"),
        category: "Components"
    },
    {
        id: "sidebar",
        name: "Sidebar",
        url: "/sidebar",
        template: require("../../src/components/sidebar/docs/sidebar-examples.html"),
        category: "Components"
    },
    {
        id: "scrollShadows",
        name: "ScrollShadows",
        url: "/scrollShadows",
        template: require("../../src/components/scrollShadows/docs/scroll-shadows-examples.html"),
        category: "Components"
    },
    {
        id: "tabs",
        name: "Tabs",
        url: "/tabs",
        template: require("../../src/components/tabs/docs/tabs-examples.html"),
        category: "Components"
    },
    {
        id: "tabsRouter",
        name: "TabsRouter",
        url: "/tabsRouter",
        template: require("../../src/components/tabs/docs/tabsRouter-examples.html"),
        category: "Components"
    },
    {
        id: "pager",
        name: "Pager",
        url: "/pager",
        template: require("../../src/components/pager/docs/pager-examples.html"),
        category: "Components"
    },
    {
        id: "toolbar",
        name: "Toolbar",
        url: "/toolbar",
        template: require("../../src/components/toolbar/docs/toolbar-examples.html"),
        category: "Components"
    },
    {
        id: "treeview",
        name: "Treeview",
        url: "/treeview",
        template: require("../../src/components/treeview/docs/treeview-examples.html"),
        category: "Components"
    },
    {
        id: "content",
        name: "Content",
        url: "/content",
        template: require("../../src/components/content/docs/content-examples.html"),
        category: "Components"
    },
    {
        id: "wizard",
        name: "Wizard",
        url: "/wizard",
        template: require("../../src/components/wizard/docs/wizard-examples.html"),
        category: "Components"
    },
    {
        id: "datetimePicker",
        name: "Datetime picker",
        url: "/datetime-picker",
        template: require("../../src/components/datetimePicker/docs/datetimePicker-examples.html"),
        category: "Components"
    },
    {
        id: "dragDrop",
        name: "Drag drop",
        url: "/drag-drop",
        template: require("../../src/components/dragdrop/docs/dragDrop-examples.html"),
        category: "Components"
    },
    {
        id: "icons",
        name: "Icons",
        url: "/icons",
        template: require("../../src/components/icon/docs/icon-examples.html"),
        category: "Components"
    },
    {
        id: "image",
        name: "Image",
        url: "/image",
        template: require("../../src/components/image/docs/image-examples.html"),
        category: "Components"
    },
    {
        id: "label",
        name: "Label",
        url: "/label",
        template: require("./templates/components/Label.html"),
        controller: "DemoLabelController as vm",
        category: "Components"
    },
    {
        id: "timepicker",
        name: "Time picker",
        url: "/timepicker",
        template: require("../../src/components/timepicker/docs/timepicker-examples.html"),
        category: "Components"
    },
    {
        id: "timeframepicker",
        name: "Timeframe picker",
        url: "/timeframepicker",
        template: require("../../src/components/timeframePicker/docs/timeframePicker-examples.html"),
        category: "Components"
    },
    {
        id: "ifShow",
        name: "If-Show",
        url: "/if-show",
        template: require("./templates/components/IfShow.html"),
        category: "Components"
    },
    {
        id: "slider",
        name: "Slider",
        url: "/slider",
        template: require("../../src/components/slider/docs/slider-examples.html"),
        category: "Components"
    },
    {
        id: "slidable",
        name: "Slidable",
        url: "/slidable",
        template: require("../../src/components/slidable/docs/slidable-examples.html"),
        category: "Components"
    },
    {
        id: "code",
        name: "Code",
        url: "/code",
        template: require("../../src/components/code/docs/code-examples.html"),
        category: "Components"
    },
    {
        id: "codeItem",
        name: "CodeItem",
        url: "/code-item",
        template: require("../../src/components/codeItem/docs/codeItem-examples.html"),
        category: "Components"
    },
    {
        id: "spreadsheet",
        name: "Spreadsheet",
        url: "/spreadsheet",
        template: require("../../src/components/spreadsheet/docs/spreadsheet-examples.html"),
        category: "Components"
    },
    {
        id: "colors-palette",
        name: "Palette (internal usage only)",
        url: "/colors-palette",
        template: require("../../src/styles/docs/colors/palette.html"),
        category: "Styles"
    },
    {
        id: "colors-semantic",
        name: "Semantic Colors",
        url: "/colors-semantic",
        template: require("../../src/styles/docs/colors/colors.html"),
        category: "Styles"
    },
    {
        id: "spacings",
        name: "Spacings",
        url: "/spacings",
        template: require("../../src/styles/docs/spacings/spacings.html"),
        category: "Styles"
    },
    {
        id: "typography",
        name: "Typography",
        url: "/typograhy",
        template: require("../../src/styles/docs/typography/typography.html"),
        category: "Styles"
    },
    {
        id: "moment",
        name: "Moment.js",
        url: "/moment",
        template: require("../../src/styles/docs/moment/moment.html"),
        category: "Styles"
    }
];
