// @solarwinds/xui/ng1/spec/components/expander/expander.atom.ts
import {by, ElementFinder, promise} from "@solarwinds/uzi/protractor";
import {Atom} from "@solarwinds/uzi";

export class ExpanderAtom extends Atom {
    static CSS_CLASS = "xui-expander";

    public isExpanded(): promise.Promise<boolean> {
        return super.getElement()
            .element(by.css(".xui-expander-body.collapse.in"))
            .isPresent();
    }

    public isContentDisplayed(): promise.Promise<boolean> {
        return super.getElement().element(by.css(".xui-expander-body")).isDisplayed();
    }

    public isContentAttachedToDOM (): promise.Promise<boolean> {
        return super.getElement().element(by.css(".xui-expander-body div")).isPresent();
    }

    public toggle(): promise.Promise<void> {
        return super.getElement().element(by.css(".xui-expander-heading-icon")).click();
    }

    public getHeading(): ElementFinder {
        return super.getElement().element(by.css(".xui-expander-heading-content"));
    }

    public getHeadingText(): promise.Promise<string> {
        return super.getElement().element(by.css(".xui-expander-heading-content")).getText();
    }
}
