// @solarwinds/xui/ng1/sdk/spec/components/menu/menu.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
const index_atom_1 = require("../index.atom");
class MenuAtom extends uzi_1.Atom {
    // Clicks on menu button
    click() {
        return this.getButtonElement().click();
    }
    // Expands menu list
    expand() {
        return this.isExpanded().then((expanded) => {
            if (!expanded) {
                this.click();
            }
        });
    }
    // Closes menu list
    close() {
        return this.isExpanded().then((expanded) => {
            if (expanded) {
                this.click();
            }
        });
    }
    // Returns whether menu list is expanded
    isExpanded() {
        return this.getMenuListElement().isDisplayed();
    }
    // Returns checkbox atom for option menu item specified by its title
    getCheckboxElemForOptionItemByTitle(title) {
        return uzi_1.Atom.findIn(index_atom_1.CheckboxAtom, super.getElement()
            .all(protractor_1.by.className("xui-menu-item"))
            .filter((elem) => {
            let checkbox = uzi_1.Atom.findIn(index_atom_1.CheckboxAtom, elem);
            return checkbox.getContent().then(function (text) {
                return text === title;
            }, function () {
                // Checkbox was not found inside the item
                return false;
            });
        })
            .get(0));
    }
    // Returns anchor element for link/action menu item specified by its title
    getAnchorElemForLinkOrActionItemByTitle(title) {
        return super.getElement()
            .all(protractor_1.by.className("xui-menu-item-link"))
            .filter((elem) => {
            return elem.getText().then(function (text) {
                return text === title;
            });
        })
            .get(0)
            .getWebElement();
    }
    // Returns anchor element for link/action menu item specified by its title containing text
    getAnchorElemForLinkOrActionItemByTitleContainingText(text) {
        return super.getElement()
            .all(protractor_1.by.className("xui-menu-item-link"))
            .filter((elem) => {
            return elem.getText().then(function (innerHtml) {
                return innerHtml.indexOf(text) !== -1;
            });
        })
            .get(0)
            .getWebElement();
    }
    // Returns if menu item element specified by its title containing text is enabled or not
    isMenuItemEnabledByTitleContainingText(text) {
        return super.getElement()
            .all(protractor_1.by.className("xui-menu-item"))
            .filter((elem) => {
            return elem.getText().then(function (innerHtml) {
                return innerHtml.indexOf(text) !== -1;
            });
        })
            .get(0)
            .getWebElement()
            .getAttribute("class").then((value) => {
            return value === null || value.indexOf("disabled") === -1;
        });
    }
    // Returns menu's button element
    getButtonElement() {
        return super.getElement().element(protractor_1.by.className("xui-menu-button")).getWebElement();
    }
    // Returns menu's list element
    getMenuListElement() {
        return super.getElement().element(protractor_1.by.className("xui-menu-list")).getWebElement();
    }
    isAppendedToBody() {
        return super.getElement().element(protractor_1.by.css(".dropdown[dropdown-append-to-body]")).isPresent() &&
            protractor_1.browser.element(protractor_1.by.css("body > .xui-menu-list.dropdown-menu")).isPresent();
    }
}
MenuAtom.CSS_CLASS = "xui-menu";
exports.MenuAtom = MenuAtom;
