// @solarwinds/xui/ng1/sdk/spec/components/progress/progress.atom.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class ProgressAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.getWidth = () => {
            return super.getElement().getCssValue("width").then((widthVal) => {
                return parseFloat(widthVal.replace("px", ""));
            });
        };
        this.getActualHeight = () => {
            return super.getElement().getSize().then((size) => {
                return size.height;
            });
        };
        this.isDisplayed = () => {
            return super.getElement().isDisplayed();
        };
        this.isProgressBarDisplayed = () => {
            return super.getElement().element(protractor_1.by.className("xui-progress__bar")).isDisplayed();
        };
        this.getMessage = () => {
            const message = super.getElement().element(protractor_1.by.className("xui-progress__message"));
            return message.getText();
        };
        this.canCancel = () => {
            return super.getElement().element(protractor_1.by.className("xui-progress__cancel")).isPresent();
        };
        this.cancelProgress = () => {
            const cancelEl = super.getElement().element(protractor_1.by.className("xui-progress__cancel"));
            return cancelEl.click();
        };
        this.waitForProgressBarToComplete = (timeout) => __awaiter(this, void 0, void 0, function* () {
            let progressBarValueWidth = 0;
            let checkInterval = 100;
            let i = 0;
            const progressBarValue = yield protractor_1.browser.findElements(protractor_1.by.className("xui-progress__bar-value")).then((array) => {
                return array[0];
            });
            while (progressBarValueWidth < 100) {
                progressBarValueWidth = yield progressBarValue.getAttribute("style").then((value) => {
                    return parseFloat(value.replace("%", "") && value.replace("width: ", ""));
                });
                protractor_1.browser.sleep(checkInterval);
                if (i > timeout / checkInterval) {
                    console.log("Quit by timeout because of these possible reasons:");
                    console.log("1. It took progress bar too long to complete");
                    console.log("2. Progress bar wasn't triggered");
                    console.log("3. Progress bar is not visible");
                    console.log("4. The set timeout is too short");
                    return;
                }
                i++;
            }
            ;
        });
    }
}
ProgressAtom.CSS_CLASS = "xui-progress";
exports.ProgressAtom = ProgressAtom;
