// @solarwinds/xui/ng1/spec/components/watchCounter/watchCounter.atom.ts
import {browser, by, element, promise} from "@solarwinds/uzi/protractor";
import {Atom} from "@solarwinds/uzi";

export class WatchCounterAtom extends Atom {
    public static CSS_CLASS = "xui-watch-counter";

    public isDisplayed = (): promise.Promise<boolean> => {
        return this.getElement().isDisplayed();
    };

    static getBadge(): WatchCounterAtom {
        return new WatchCounterAtom(element(by.className("xui-watch-counter__highlight")));
    }

    public getTextColor() {
        return super.getElement().getCssValue("color");
    }

    public getBackgroundColor() {
        return super.getElement().getCssValue("background-color");
    }

    public mouseHover = (): promise.Promise<void> => {
        return browser.actions()
            .mouseMove(this.getElement().getWebElement())
            .perform();
    };

    public getBorderStyle() {
        return super.getElement().getCssValue("outline");
    }

    public getTextNum() {
        return super.getElement().getText();
    }
}
