// @solarwinds/xui/ng1/sdk/spec/components/accordion/accordion.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
class AccordionAtom extends uzi_1.Atom {
    static findIn(atomClass, parentElement, index) {
        if (!parentElement) {
            throw "Accordion atom doesn't support global search yet";
        }
        let componentRootElement;
        if (index !== undefined) {
            componentRootElement = parentElement.all(protractor_1.by.xpath(".//uib-accordion")).get(index);
        }
        else {
            componentRootElement = parentElement.element(protractor_1.by.xpath(".//uib-accordion"));
        }
        return new atomClass(componentRootElement);
    }
    getGroup(groupName) {
        return new AccordionGroup(this.getElement().element(protractor_1.by.xpath("./div/div[@group-name='" + groupName + "']")));
    }
}
exports.AccordionAtom = AccordionAtom;
class AccordionGroup extends uzi_1.Atom {
    isExpanded() {
        return super.getElement().element(protractor_1.by.css(".xui-icon.xui-icon-triangle-down")).isPresent();
    }
    isCollapsed() {
        return super.getElement().element(protractor_1.by.css(".xui-icon.xui-icon-triangle-right")).isPresent();
    }
    toggleExpand() {
        return this.getHeaderElement().click();
    }
    getHeaderElement() {
        return super.getElement().element(protractor_1.by.css(".accordion-toggle"));
    }
}
exports.AccordionGroup = AccordionGroup;
