// @solarwinds/xui/sdk/spec/components/actionDrawer/actionDrawer.atom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const protractor_1 = require("@solarwinds/uzi/protractor");
const uzi_1 = require("@solarwinds/uzi");
const specHelpers_1 = require("../specHelpers");
class ActionDrawerAtom extends uzi_1.Atom {
    constructor() {
        super(...arguments);
        this.moveMouseToTarget = (atom) => {
            return specHelpers_1.default.mouseOver(atom);
        };
    }
    isTooltipDisplayed() {
        return this.getTooltip().isPresent();
    }
    getTooltipAttribute(attribute) {
        return this.getTooltip().getAttribute(attribute);
    }
    getMenuIconAttribute(iconType, attribute) {
        return super.getElement()
            .element(protractor_1.by.css(`.${ActionDrawerAtom.CSS_CLASS} [icon=${iconType}]`))
            .getAttribute(attribute);
    }
    clickMenuIcon(iconType) {
        return super.getElement()
            .element(protractor_1.by.css(`.${ActionDrawerAtom.CSS_CLASS} [icon=${iconType}]`))
            .click();
    }
    moveMouseToMenuIcon(iconType) {
        return protractor_1.browser.actions().mouseMove(super.getElement()
            .element(protractor_1.by.css(`.${ActionDrawerAtom.CSS_CLASS} i[icon=${iconType}]`)))
            .perform();
    }
    getMenuDirection() {
        return super.getElement()
            .element(protractor_1.by.className("xui-action-drawer__actions")).getCssValue("transform");
    }
    getToastContainer() {
        return protractor_1.browser.element(protractor_1.by.id("toast-container")).isPresent();
    }
    getTooltip() {
        return protractor_1.browser.element(protractor_1.by.className("action-drawer-tooltip"));
    }
}
ActionDrawerAtom.CSS_CLASS = "xui-action-drawer";
exports.ActionDrawerAtom = ActionDrawerAtom;
